RFC: Support Dockerfiles (#173)
buildpacks/rfcs: text/0000-dockerfiles.md

# Meta
[meta]: #meta
- Name: Support Dockerfiles
- Start Date: 2021-06-30
- Author(s): sclevine
- RFC Pull Request: (leave blank)
- CNB Pull Request: (leave blank)
- CNB Issue: (leave blank)
- Supersedes: [RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md), [RFC#167](https://github.com/buildpacks/rfcs/pull/167)
- Depends on: [RFC#172](https://github.com/buildpacks/rfcs/pull/172)

# Summary
[summary]: #summary

This RFC introduces functionality for customizing base images, as an alternative to stackpacks.

# Motivation
[motivation]: #motivation

Relying on Dockerfiles for base image generation and manipulation allows us to apply buildpacks only to the problem that they solve best: managing application runtimes and dependencies.

# What it is
[what-it-is]: #what-it-is

This RFC proposes that we replace stackpacks with multi-purpose build-time and runtime Dockerfiles.

# How it Works
[how-it-works]: #how-it-works

> Note: kaniko, buildah, BuildKit, the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 

Reviewer: Should this be entirely up to the platform or should we provide a new lifecycle phase(s) to support this behavior? I think we probably need to do the latter if we want consistent application of the  `io.buildpacks.rebasable=false` or `io.buildpacks.sbom` labels.

Author: Agree, intent was just to say that platforms can do what they'd like to apply the Dockerfiles.

Author: (closing comments so that PR page loading stops timing out)

### App-specified Dockerfiles

> A buildpack app may have a build.Dockerfile and/or run.Dockerfile in its app directory. A run.Dockerfile is applied to the selected runtime base image after the detection phase. A build.Dockerfile is applied to the build-time base image before the detection phase.

Reviewer 2: As a developer, I like giving all this power to app devs, but if I put on my Operations hat, I don't want my devs to have that much power. How would this be controlled? Or is that up to the platform implementation? 

Possibly also an implementation detail, but I think that allowing this level of customization should be something called out very clearly in the output/logs of the process applying these changes. As a buildpack dev, I worry that someone will make a change that breaks my buildpack, then turn around and point a finger saying my buildpack is broken. If it's not extremely clear what customizations are made, it'll be difficult to debug issues like this.

Author: Agree -- platform should be able to opt-out of providing this feature, and the Dockerfile build output should appear in logs.

> A buildpack app may have a build.Dockerfile and/or run.Dockerfile in its app directory. A run.Dockerfile is applied to the selected runtime base image after the detection phase. A build.Dockerfile is applied to the build-time base image before the detection phase.

Author: Will clarify. Intent is to propose that a `build.Dockerfile` and `run.Dockerfile` be placed adjacent to project.toml and applied automatically on `pack build`.

As with stackpacks, BuildKit (or the older docker daemon) could be used in place of kaniko for local builds.

> A buildpack app may have a build.Dockerfile and/or run.Dockerfile in its app directory. A run.Dockerfile is applied to the selected runtime base image after the detection phase. A build.Dockerfile is applied to the build-time base image before the detection phase.

Reviewer 3: > Will clarify. Intent is to propose that a `build.Dockerfile` and `run.Dockerfile` be placed adjacent to project.toml and applied automatically on `pack build`.

What will happen during the execution of the command "packer build" if no `*.Dockerfile` files exist within the developer's project ?

> Both Dockerfiles must accept `base_image` and `build_id` args. The `base_image` arg allows the lifecycle to specify the original base image. The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`.  

Reviewer 2: Can you expand on `build_id`? How is that working to bust the cache?

Author: It changes every time, so using it in a layer invalidates that layer and all subsequent layers. Will add this to the RFC.

> Both Dockerfiles must accept `base_image` and `build_id` args. The `base_image` arg allows the lifecycle to specify the original base image. The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`.  

Reviewer 3: Will it be possible to override the `base_image` ARG using the `pack build` command or any other `docker | podman | buildah build` command ?
example:

```
pack build -a build_base_image=docker.io/swnodrop/buildpack-jdk:1.1 .... -a run_base_image=....`
```

Author: Yes, but not directly. You can override the build-time base image by selecting a different builder (which is important, because the buildpack has to be present on the image). You can override the run image by selecting a different run image (`pack build --run-image`). This has the same outcome.

Reviewer 3: If I follow you here, that means that the `base_image` should be hard coded within the `builder.toml | yaml` file and that we will pickup for the build a different builder ?

Author: The builder is always used as the build-time base image. It may be extended by a build.Dockerfile, but it cannot be replaced (given that it contains the buildpacks and lifecycle).

Reviewer 3: So, how will it be possible to extend the builder definition to specify a `build.dockerfile` ? Will then this `build.Dockerfile` be used as first step to build/push a new image ?

Author: The builder would not specify a build.Dockerfile. Instead, a build.Dockerfile could be applied to the builder before a build (to generate a new builder) or during a build.

> Both Dockerfiles must accept `base_image` and `build_id` args. The `base_image` arg allows the lifecycle to specify the original base image. The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`.  

Reviewer 3: How can the developer be aware what the `build_id` value is, what the layer or cache contains ?

Author: The value is arbitrary, just different on each build. The developer only cares about `build_id` when they're creating a Dockerfile that uses `build_id`.

> A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of app-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before `run.Dockerfile` is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.

Reviewer 4: I assume this will be tue for`/cnb/lifecycle/rebaser` also?

What is the behavior when trying to rebase an image that has `io.buildpacks.rebasable=false`? Error or no-op?

Author: > I assume this will be true for/cnb/lifecycle/rebaser also?

Definitely

> What is the behavior when trying to rebase an image that has io.buildpacks.rebasable=false? Error or no-op?

Error -- the user should know that the rebase did not succeed, so they can rebuild.

Reviewer 2: I wonder about the concept of being "rebasable" and if users will understand that. I could see someone just marking rebasable = true because it sounds better or having users just always add the `-f` flag (cause what could go wrong with that?), and not really understand what it means to be rebasable or ABI compatible. 

It makes me wonder what will happen if you force it to true & try to rebase? Are you going to get clear and obvious errors or are you going to get very subtle and difficult to debug errors? Best case, maybe you get shared library loading problems as the app starts up and you know something is horribly wrong? Maybe worst case, things fail later as your code tries to do certain things and load additional libraries (hopefully soon enough after the rebase that you can connect the dots)? 

I'm not sure there's much you can do about it. Maybe you could have a safety check like force the user to confirm when they have a Dockerfile that contains rebasable=true? That still wouldn't solve it though. 

Maybe we just need a note to flag the importance of documenting this label so users know when it's safe to mark it as true.

Reviewer 3: > I wonder about the concept of being "rebasable" and if users will understand that

Such a property `io.buildpacks.rebasable=true` will certainly create confusion as the end users will not understand the reason why the image is not rebasable (Is it because, for licensing , security, business or technical reasons it cannot be rebased ?). What should they do then when the `pack rebase` will report such a message `image not rebasable` (rebuild, contact their `builder`s team, security advisor, ops team,...) ?

Author: Aside from clear error messages and warnings, I think the alternative would be to permanently disable rebase any time a Dockerfile is used to extend an existing stack. I prefer clear error messages and warnings.

Reviewer 2: My $0.02, we should optimize for the common use case. We know that most users will want to use this feature add some additional libraries to the stack. The common request is "I need libfoo or foo-cli to compile/run my app, how do I install that?". Since that's likely what most users will do, we should optimize for that case. If it's generally going to produce a rebasable container, then it seems reasonable to leave it up to the user (who, I believe, will very likely mark it true without a lot of thought). If it's not generally going to produce a rebasable container, we should just always mark the container as not rebasable permanently. If it's complicated enough that we can't clearly tell if the happy path will produce rebasable containers, then we should probably mark an extended container as permanently not rebasable.

Author: > who, I believe, will very likely mark it true without a lot of thought

As specified, it would be reset to false before any subsequent Dockerfile is applied, and would not need to be present in examples. The user would have to look up the label in the docs and intentionally add it to enable rebasing.

Reviewer 2: Yes, I think that's exactly what'll happen. StackOverflow, copy & paste, boom. They set it to rebasable without understanding anything (or at best very little) about what that means. For what it's worth though, I think the same will happen with the rebase command, users will add `--force` because it "works" not really understanding the nuance.

I am a pessimist regarding stuff like this, 10+ years in support will do that to you though :) I'm OK to resolve this comment if you'd prefer to handle this through docs/errors & warnings. You either allow folks to shoot themselves in the foot and potentially mark something rebasable that's not, or it's never rebasable when you use app Dockerfiles. I can get why you want to allow it for responsible users, rebasing is a nice feature.

Author: I think the new hooks model partially mitigates this by making the Dockerfiles a builder-level concern.

Author: (closing comments so that PR page loading stops timing out)

### Platform-specified Dockerfiles

> The same Dockerfiles may be used to create new base images or modify existing base images outside of the app build process. For both app-specified and base-image-modifying Dockerfiles, any specified labels override existing values.

Reviewer 3: What do you mean here `outside of the app build process` (that the `dockerfiles` part of a buildpack `builspack/<buildpackA/bin/` folder could also be used by a developer on their laptop to manually execute a `docker build` ?

Reviewer 2: Two things I picked up from the working group discussion yesterday about this:

1. the Dockerfiles addition here is for app devs and for stack creators, it's not attempting to address buildpack authors. There is no mechanism for a buildpack author to provide a Dockerfile or to indicate that certain packages would need to be installed (I think #172 has some capabilities that a buildpack author could use to indicate requirements from the stack, that that #172 would install them but it allows you to indicate what is necessary from a stack to use your buildpacks).

2. "outside of the app build process" would refer to someone wanting to create a stack, so perhaps using `pack` cli to `pack stack create` or something like that (obviously other platforms could support similar capabilities).

Reviewer 3: > 1. There is no mechanism for a buildpack author to provide a Dockerfile

Who will then provide such `run.Dockerfile` and `build.Dockerfile` files ?

Reviewer 2: The two methods outlined in the RFC here are 1.) app devs and 2.) stack creators.

For apps devs, this RFC says `A buildpack app may have a build.Dockerfile and/or run.Dockerfile in its app directory.` so those would sit in a similar location as project.toml.

For stack creators, the location isn't tied to anything. You could manage them however you like. You'd then end up using a tool like `pack` to create the stack based on your customized docker files. Maybe `pack stack create`, I'm not sure if that was set in the RFC.

Reviewer 5: @dmikusa-pivotal this thread is a bit old, but I think having hooks interact with the build plan now provides a way for buildpack authors to get things they need from hooks. A buildpack could indicate that it `requires` a dependency and trust that a hook will be there to `provide` it.

> Dockerfiles that are used to create a stack must create a `/cnb/stack/genpkgs` executable that outputs a [CycloneDX](https://cyclonedx.org)-formatted list of packages in the image with PURL IDs when invoked. This executable is executed after any run.Dockerfile or build.Dockerfile is applied, and the output replaces the label `io.buildpacks.sbom`. This label doubles as a Software Bill-of-Materials for the base image. In the future, this label will serve as a starting point for the application SBoM.

Reviewer 4: why isn't `/cnb/stack/genpkgs` optional?

Reviewer 4: If the "stack" concept is being removed in https://github.com/buildpacks/rfcs/pull/172 is this the right name for this dir?

Author: > why isn't /cnb/stack/genpkgs optional?

It's only required when creating a base image, not for app-specific Dockerfiles. I feel like it should be required so that changes from app-specific Dockerfiles (or from the base-image-creating Dockerfile) aren't silently dropped from the SBoM by accident. To opt-out, I'd like the base image author to explicitly creates `genpkgs` and make it no-op, so they understand that the SBoM will be inaccurate (which seems extremely dangerous, given that a CycloneDX SBoM strongly implies that the metadata can be used for vulnerability scanning).

> If the "stack" concept is being removed in #172 is this the right name for this dir?

Definitely not, good catch. I should also drop the word "stack" from this RFC.

Author: Moved to `/cnb/image/genpkgs`

Reviewer 3: > author to explicitly creates `genpkgs` and

How can we create such `genpkgs` executable ? Where it it documented ? Example available ?

Author: It's an executable that run on the base image and outputs CycloneDX JSON to stdout. Happy to add more detail.

Reviewer 3: > outputs CycloneDX JSON to stdout

Will this executable called during the build step (executed by lifecycle creator) and will output the JSON file containing the packages of the `build` but also the `run` image ? 

How will it be possible next to use such CycloneDX JSON file(s) to document the content of an image ?
Example
```
pack inspect buildpack{run|build}image_name
-->
...
packages:
   - package1
   - package2
   - package3
   - ....
```

Author: > Will this executable called during the build step (executed by lifecycle creator) and will output the JSON file containing the packages of the build but also the run image ?

It's not executed during the build phase (i.e., after all Dockerfiles have been applied), because the base image can't be modified by buildpacks. Buildpacks will be able to contribute their own entries to the SBoM via the buildpack BoM interface. See: https://github.com/buildpacks/rfcs/pull/166

Author: > How will it be possible next to use such CycloneDX JSON file(s) to document the content of an image ?

The CycloneDX project/ecosystem has rich tooling for this. It would be great if `pack inspect` could take advantage of that tooling to introspect all types of images (app images, base images, buildpack images).

Reviewer 5: Marking as resolved given that we're leaving SBOM generation up to the platform (see https://github.com/buildpacks/spec/pull/298#issuecomment-1099543096).

> Dockerfiles that are used to create a base image must create a `/cnb/image/genpkgs` executable that outputs a [CycloneDX](https://cyclonedx.org)-formatted list of packages in the image with PURL IDs when invoked. This executable is executed after any run.Dockerfile or build.Dockerfile is applied, and the output replaces the label `io.buildpacks.sbom`. This label doubles as a Software Bill-of-Materials for the base image. In the future, this label will serve as a starting point for the application SBoM.

Reviewer 2: Pardon my ignorance in the realm of bill-of-materials, but why not just use a flat file (i.e. one package per line) or simple JSON format? Writing and parsing would seem to be easier that way, reducing the difficulty of creating `genpkgs`. 

What's the driver for CycloneDX? Is there a compatibility concern with external tools? Is the more complicated format required to represent the information? Is there a general tool that one could just drop in here as `genpkgs` and make it work? 

My thought is that having to write a tool for `genpkgs` and use the CycloneDX format is a hurdle for someone wanting to make their own base images.

Reviewer 3: > My thought is that having to write a tool for `genpkgs` and use the CycloneDX format is a hurdle for someone wanting to make their own base images.

I fully agree. Otherwise, buildpack should provide a tool, documentation, ... to generate it automatically using list of packages generated as yaml text file

Author: A few points:
1. It's important that genpkgs be an executable, so that the SBoM can be updated automatically when the base image is extended (including with app-specified Dockerfiles)
2. It's important that the format be CycloneDX, given https://github.com/buildpacks/rfcs/pull/166. The goal is to reliably produce an SBoM in an industry-standard and security-focused format.
3. CycloneDX appears to have robust tooling for both security scanning and SBoM generation, e.g., https://github.com/CycloneDX/cyclonedx-maven-plugin
4. I agree that the project could provide genpkgs for common linux distros -- `pack create-stack` could even have a flag to add it automatically

(Will add these points to the RFC.)

### Examples

run.Dockerfile used to create a runtime base image:

```
ARG base_image
FROM ${base_image}
ARG build_id=0

LABEL io.buildpacks.image.distro=ubuntu
LABEL io.buildpacks.image.version=18.04
LABEL io.buildpacks.rebasable=true

ENV CNB_USER_ID=1234
ENV CNB_GROUP_ID=1235

RUN groupadd cnb --gid ${CNB_GROUP_ID} && \
  useradd --uid ${CNB_USER_ID} --gid ${CNB_GROUP_ID} -m -s /bin/bash cnb

USER ${CNB_USER_ID}:${CNB_GROUP_ID}

COPY genpkgs /cnb/image/genpkgs
```

run.Dockerfile present in an app directory that always installs the latest version of curl:
```
ARG base_image
FROM ${base_image}
ARG build_id=0

> LABEL io.buildpacks.rebasable=true

Author: Typo: should be false

RUN echo ${build_id}

RUN apt-get update && apt-get install -y curl && rm -rf /var/lib/apt/lists/*
```

Unsafe run.Dockerfile present in an app directory:
```
ARG base_image
FROM ${base_image}
ARG build_id=0

LABEL io.buildpacks.rebasable=false

RUN curl -L https://example.com/mypkg-install | sh # installs to /opt
```

# Drawbacks
[drawbacks]: #drawbacks

- Involves breaking changes.
- Buildpacks cannot install OS packages directly, only select runtime base images.

# Alternatives
[alternatives]: #alternatives

- Stackpacks

# Unresolved Questions

> [unresolved-questions]: #unresolved-questions

Reviewer 3: Will it be possible to execute a `run.Dockerfile` or `build.Dockerfile` with a `root` user to install additional packages (.eg. maven, jdk, ..) as such option was included within the RFC69 which is now closed ?

Reviewer 2: My understanding from the working group discussion yesterday is that the dockerfiles execute as root, they can basically do anything to the container.

Reviewer 3: > My understanding from the working group discussion yesterday is that the dockerfiles execute as root, they can basically do anything to the container.
and

> Only the Dockerfiles can run as root. This proposal does not allow buildpacks to run as root.

Can you elaborate please. Which application, system will then execute the command to build as root the Dockerfiles ?

Reviewer 2: I don't recall that being discussed in the meeting. I would have to defer to @sclevine

Author: The platform would likely orchestrate new lifecycle phases that apply the Dockerfiles. This RFC does not (yet) propose an implementation, just UX.

Reviewer: typically the instrucitons in a dockerfile would run as the `USER` set on the base image. Would we expect Dockerfiles to explicilty set `USER root` and then switch back to `USER cnb`?

Author: As proposed, yes, but I could definitely be convinced otherwise. We could restore the original UID/GID if the final values are 0:0.

- Should `genpkgs` be part of this proposal? Opinion: Yes, otherwise it's difficult to maintain a valid SBoM.

> 

Reviewer 2: Does this weaken the security stance of buildpacks? (My opinion: yes) You can literally do anything in a Dockerfile. 

I feel like stack packs, while maybe not the answer, did put more controls around what you could do. The stack packs were buildpacks, so presumably (IDK exactly as it wasn't implemented), an Operations team could control which stack packs are available to their users, thereby limiting what their users could install or modify to some degree. 

There's also a stronger guarantee that a stack pack would produce a legitimate BOM, since Operations teams can control what stack packs are available (and audit them). With this proposal, someone could make a Dockerfile that installs something malicious and set `genpkgs` to be a copy of the `true` binary or some other no-op binary. The proposal essentially trusts the creator of the Dockerfile to be honest and report what they've installed, which I don't think is merited givn that this proposal would allow any app dev to customize what's install with a Dockerfile.

Reviewer 5: > Operations teams can control what stack packs are available (and audit them)

I don't have any answers here, but this seems like a fairly important point to linger on.

Reviewer 5: I think this concern is pretty well addressed by including the hooks on the builder. If the app-specified Dockerfile hook (above example) is not present, then the developer cannot provide a Dockerfile. And I suppose that trusting hook authors to provide a valid BOM is the same as trusting buildpack authors to provide a valid BOM.

In the case of the app-specified Dockerfile, the hook author could apply some label to the image so that consumers would know to approach the BOM with some suspicion. I wonder if there's value in having that spec'd ('io.buildpacks.extended'?). It's just a thought.

Reviewer 6: I like the idea of a label to easily identify images that have had any hooks executed.

Reviewer 5: Also `/cnb/image/genpkgs` (because it is executed after Dockerfiles have been applied) is helping to trust the BOM.

Reviewer 5: @samj1912 and I have recently been discussing having the platform inject `genpgks`. That may help here...

Reviewer 6: > @samj1912 and I have recently been discussing having the platform inject `genpgks`. That may help here...

How would that work? I don't think a mount would work, since the hook could discover and overwrite it. Since the hook could do anything, I'm not even sure _running_ a program like `genpgks` is safe even if the binary was untouched. I would think the `genpgks` would almost have to be executed in a new container against the extended image to truly be accurate.

That said - these hooks are likely authored and for sure approved/distributed by the stack author. Just because `Dockerfile`s are used, doesn't mean the stack distributor has to extend that control to buildpack authors or the app author. They can create a much more simple constraint (like defining an `Aptfile` to read package names).

Reviewer 5: > I would think the genpgks would almost have to be executed in a new container against the extended image to truly be accurate.

That's basically what we're proposing - please see https://github.com/buildpacks/rfcs/pull/173/files#r794608075 for the download I got from @samj1912 and @sclevine .

# Spec. Changes (OPTIONAL)
[spec-changes]: #spec-changes

> This RFC requires extensive changes to all specifications.

Reviewer 4: nods

# Meta
[meta]: #meta
- Name: Support Dockerfiles
- Start Date: 2021-06-30
- Author(s): sclevine
- RFC Pull Request: (leave blank)
- CNB Pull Request: (leave blank)
- CNB Issue: (leave blank)
- Supersedes: [RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md), [RFC#167](https://github.com/buildpacks/rfcs/pull/167)
- Depends on: [RFC#172](https://github.com/buildpacks/rfcs/pull/172)

# Summary
[summary]: #summary

This RFC introduces functionality for customizing base images, as an alternative to stackpacks.

# Motivation
[motivation]: #motivation

Relying on Dockerfiles for base image generation and manipulation allows us to apply buildpacks only to the problem that they solve best: managing application runtimes and dependencies.

# What it is
[what-it-is]: #what-it-is

This RFC proposes that we replace stackpacks with multi-purpose build-time and runtime Dockerfiles.

# How it Works
[how-it-works]: #how-it-works

Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 

> ### Builder-specified Dockerfiles

Reviewer 7: I really like these additions you have made!

A builder may specify any number of executable "hooks" in `/cnb/hooks.d/`.
Hook files in this directory are executed in the context of app directory and must output Dockerfiles that are applied to the build-time and runtime base images before an image is built.

If a hook exits with a non-zero status value, the build fails. If a hook exits with a zero status value and no output, the hook is ignored. Directories and non-executable files are ignored.

> Each executable must be in the format `/cnb/hooks.d/<name>.(build.|run.|)<format>`, where build, run, or empty specify the build-time base image, runtime base image, or both bash images, respectively. The only valid format is `Dockerfile`, although support for, e.g. LLB JSON, could be added in the future. 

Reviewer 8: As I envision this hook directory listing, not considering some of the good thoughts @cmoulliard has brought up, I would prefer to group the files by phase first, then order, and lastly format. 

It seems easier to reason about the order of operations:

#### Current

```
./hooks.d/1_step1.Dockerfile
./hooks.d/1_step1.build.Dockerfile
./hooks.d/1_step1.run.Dockerfile
./hooks.d/2_step2.Dockerfile
./hooks.d/2_step2.build.Dockerfile
./hooks.d/3_step3.build.Dockerfile
```

#### Suggested

```
./hooks.d/1_step1.Dockerfile
./hooks.d/2_step2.Dockerfile
./hooks.d/build.1_step1.Dockerfile
./hooks.d/build.2_step2.Dockerfile
./hooks.d/build.3_step3.Dockerfile
./hooks.d/run.1_step1.Dockerfile
```

Author: With the order proposed in the RFC, the list order matches the order of application.
With the order you recommend, the "both" hooks may appear out-of-order. This seems more confusing to me.

Reviewer 5: Now that hooks -> extensions are part of an order, and have a `/cnb/extensions` directory, this is moot. I'm going to mark this as resolved.

A runtime Dockerfile is applied to the selected runtime base image after the detection phase.
A build-time Dockerfile is applied to the build-time base image before the detection phase.

Both Dockerfiles must accept `base_image` and `build_id` args.
The `base_image` arg allows the lifecycle to specify the original base image.
The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`. When the `$build_id` arg is referenced in a `RUN` instruction, all subsequent layerrs will be rebuilt on the next build (as the value will change).

> A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of app-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before `run.Dockerfile` is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.

Author: ```suggestion
A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before `run.Dockerfile` is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
```

> A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before `run.Dockerfile` is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.

Author: ```suggestion
A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before a runtime Dockerfile is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
```

# Motivation
[motivation]: #motivation

Relying on Dockerfiles for base image generation and manipulation allows us to apply buildpacks only to the problem that they solve best: managing application runtimes and dependencies.

# What it is
[what-it-is]: #what-it-is

This RFC proposes that we replace stackpacks with multi-purpose build-time and runtime Dockerfiles.

# How it Works
[how-it-works]: #how-it-works

Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 

### Builder-specified Dockerfiles

A builder image may include any number of "hook" files in `/cnb/hooks.d/`.
The `pack create-builder` command is used to copy hooks into the builder image via `builder.toml`. E.g.,
```toml
[[hooks]]
name = "app-hook"
path = "./myhook"
```
(Note: this format should not be considered final. Additional fields will be required to mark the format, target, etc. of the hook. These may be defined in subsequent spec PRs / sub-team RFCs.)

Each hook file path must be in the format `/cnb/hooks.d/<name>.(build.|run.|)<format>(.out|)`, where:

> 1. `build`, `run`, or empty specify the build-time base image, runtime base image, or both bash images, respectively.

Reviewer 2: ```suggestion
1. `build`, `run`, or empty specify the build-time base image, runtime base image, or both base images, respectively.
```

2. The only valid format is `Dockerfile`, although support for, e.g. LLB JSON, could be added in the future.
3. If the `.out` suffix is present and the file is executable, the hook is executed in the context of the app directory, and its output must match `<format>`.
4. If the `.out` suffix is not present, the contents of the file must match `<format>`.

> Hook files are evaluated (either read or executed) and the resulting Dockerfiles are applied to the build-time and runtime base images in lexographical order by `<name>`.

Reviewer 3: Can you please clarify how and when the hooks will be evaluated and/or executed ? As the hook creates images (build and/or run), should we then be executed by another Buildpack step/phase (e.g pre-build, hook, ...) before the step `build` is executed ? Otherwise, it will not be possible to use the new `built` image to by example execute `maven package` as maven is not yet part of the image coming from the hooked dockerfile.

Will it be possible to pass a `build` TOML file containing METADATA to be used by the `hook` to generate the final Dockerfile(s) (e.g. MAVEN version ==> translated by the hook to the command able to install the RPM packages) ?

Author: > Can you please clarify how and when the hooks will be evaluated and/or executed ? As the hook creates images (build and/or run), should we then be executed by another Buildpack step/phase (e.g pre-build, hook, ...) before the step build is executed ?

This is stated a few lines below:

> A Dockerfile intended for the runtime base image is applied after the detection phase.
> A Dockerfile intended for the build-time base image is applied before the detection phase.

Detection happens before build. The runtime Dockerfile can be applied at any time before export, because the resulting image isn't used until that phase.

> Will it be possible to pass a build TOML file containing METADATA to be used by the hook to generate the final Dockerfile(s) (e.g. MAVEN version ==> translated by the hook to the command able to install the RPM packages) ?

If the TOML file is in the app directory, then yes, as the hook will execute in the context of the app directory.

Reviewer 3: > A Dockerfile intended for the runtime base image is applied after the detection phase.
> A Dockerfile intended for the build-time base image is applied before the detection phase.

Which inputs parameters will then be passed to the `pre-detect` or ` post-detect` phase ? Why such a question --> To be able to achieve the following scenario: During the execution of the pre or post detect phase, I would like to extract from some METADATA the version of the maven RPM to be installed or JDK requested (java 8, 11, ...) like also if JDK or JRE should be installed. Such METADATA will then be used to set some ENV VAR for the Dockerfile to be executed by kaniko, .... tool.

Author: > Which inputs parameters will then be passed to the pre-detect or post-detect phase ?

Executable hooks are run with the app directory set as the working directory, but take no other arguments (aside from env vars set on the base images).

Dockerfiles take the specified ARGS, and inherit env vars from their base images.

> I would like to extract from some METADATA the version of the maven RPM to be installed or JDK requested (java 8, 11, ...) like also if JDK or JRE should be installed.

Where does this metadata come from?

Reviewer 2: @sclevine - I was a little unclear on this point.

>A Dockerfile intended for the runtime base image is applied after the detection phase.
>A Dockerfile intended for the build-time base image is applied before the detection phase.

It says when the Dockerfiles generated will be applied, but I don't see an indication of when the hook will be executed (if it's executable).

For example, it says `A Dockerfile intended for the runtime base image is applied after the detection phase`, but that doesn't necessarily mean that the hook runs after detection. Based on the description, the hook could run prior to detection and the output be held until after.

Is this intentional to allow implementors more room to work? If not, could we clarify the time when the hooks are executed as well?

Reviewer 2: @sclevine - One more question on the execution of hooks.

The docs say that the hook will execute with the app dir set as the working directory and that it does not receive any additional arguments.

For clarification:

1. Would the hook have access to `/cnb/buildpacks/**` in the builder image? If it wanted to, for example, retrieve assets included with buildpacks? 
2. For either `build` or `run`, would there be anything under `/layers/**` yet? Wondering how this would interact with cached layers? Also, because of the timing specified here `The runtime Dockerfile can be applied at any time before export, because the resulting image isn't used until that phase.`, what you could expect? Did the buildpacks run and populate the layers yet? or even if they did would those files be exposed? 

Thanks

Author: > could we clarify the time when the hooks are executed as well?

Given:
> Executable hook files must not modify the app directory when executed and may be executed in parallel.
And:
> A Dockerfile intended for the build-time base image is applied before the detection phase.

Build-time hooks could be executed at any point before the Dockerfiles need to be applied (in parallel with buildpack detection would be ideal for performance, but the platform may choose).

It's worth clarifying that runtime hooks must be executed before the build phase (or on a copy of the original source code). I'll add this.

> Is this intentional to allow implementors more room to work? 

I'm trying not to over-specify, given that platforms may want to use their own tooling to apply the hooks, and given that the RFC will be followed by a spec PR. However, I think it's better if the RFC is clear when flexibility is allowed, so I'm happy to add more detail.

> Would the hook have access to /cnb/buildpacks/** in the builder image? If it wanted to, for example, retrieve assets included with buildpacks?

The build-time Dockerfiles run as root on the builder image, so nothing would prevent this. Anything is allowed that doesn't result in a non-compliant builder.

> For either build or run, would there be anything under /layers/** yet? Wondering how this would interact with cached layers?

`/layers` is generally a volume during the build process (so any changes would have no effect) and is generally empty in the runtime base image (so the Dockerfiles probably shouldn't create directories that conflict with buildpack directories).

> Also, because of the timing specified here The runtime Dockerfile can be applied at any time before export, because the resulting image isn't used until that phase., what you could expect? Did the buildpacks run and populate the layers yet? or even if they did would those files be exposed?

Layers are not populated on the runtime base image until the export phase.

Reviewer 2: OK, sounds good. Thanks for clarifying!

#### Example: App-specified Dockerfile Hook

This example hook would allow an app to provide runtime and build-time base image extensions as "run.Dockerfile" and "build.Dockerfile."
The app developer can decide whether the extensions are rebasable.

##### `/cnb/hooks.d/app.build.Dockerfile.out`
```
#!/bin/sh
cat build.Dockerfile
```
##### `/cnb/hooks.d/app.run.Dockerfile.out`
```
#!/bin/sh
cat run.Dockerfile
```

#### Example: RPM Dockerfile Hook

This example hook would allow a builder to install RPMs for each language runtime.

Note: The Dockerfiles referenced must disable rebasing, and build times will be slower compared to buildpack-provided runtimes.

##### `/cnb/hooks.d/app.Dockerfile.out`
```
#!/bin/sh
[[ -f Gemfile.lock ]] && cat /cnb/hooks.d/app.Dockerfile.d/Dockerfile-ruby
[[ -f package.json ]] && cat /cnb/hooks.d/app.Dockerfile.d/Dockerfile-node
```

> ### Platform-specified Dockerfiles

Reviewer 3: Will the platform-specified Dockerfiles be defined part of the builder.toml ?

Author: No, the section probably needs a better name.

These Dockerfiles would be inputs to `pack create-stack` (or whatever we call the helper command that will be used to create buildpack-compatible base images).

Reviewer 3: > These Dockerfiles would be inputs to `pack create-stack` (or whatever we call the helper command that will be used to create buildpack-compatible base images).

Ideally they should be defined part of the builder.toml file as URI or passed as command parameter(s) pointing to the Dockerfiles

Author: > Ideally they should be defined part of the builder.toml file as URI or passed as command parameter(s) pointing to the Dockerfiles
I think you're confusing this section with the section above. A better title for this section would be something like "Using Dockerfiles to Extend Base Images." The point this that the same Dockerfiles used for extending builds could also be used to create/extend base images.

# How it Works
[how-it-works]: #how-it-works

Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 

### Dynamically-applied Dockerfiles

A builder image may include any number of "hook" directories in `/cnb/hooks/`.

Hooks are similar to buildpacks: they have a `/bin/build` and `/bin/detect` executable.
However, instead of a `buildpack.toml` file, hooks have a `hook.toml` file:
```toml
api = "<buildpack API version>"

[hook]
id = "<hook ID>"
name = "<hook name>"
version = "<hook version>"
homepage = "<hook homepage>"
description = "<hook description>"
keywords = [ "<string>" ]

[[hook.licenses]]
type = "<string>"
uri = "<uri>"
```

Hooks may be packaged and examined similar to buildpacks, but with analogous `pack hook` subcommands.

Other pack CLI commands, such as `pack builder create`, will be extended to include support for hooks.

Unlike buildpacks,
- Hooks must not be included in a meta-buildpacks
- Hooks must not have `order`/`group` definitions in `hook.toml`

Hooks participate in the buildpack detection process, with the same interface for `/bin/detect`.
However,

> - `/bin/detect` is optional for hooks, and they are assumed to pass detection when it is not presetn.

Reviewer 9: What is the detect interface for hooks like?

How do they specify provides?

Author: Line 65:
> [...] with the same interface for `/bin/detect`.

Reviewer 10: Can you please provide a detailed description? Like what `/bin/detect` does? What it detects? Binary?

Reviewer 11: > Can you please provide a detailed description? Like what `/bin/detect` does? What it detects? Binary?

`/bin/detect` is here to say whether the hook should be applied or not. How is this decided is up to you. You can also omit it and then the hook is executed every time. Take a look at the same script in buildpacks, it works the same way basically.

Reviewer 5: ```suggestion
- `/bin/detect` is optional for hooks, and they are assumed to pass detection when it is not present. Just like with buildpacks, a /bin/detect that exits with a 0 exit code passes detection, and fails otherwise.
```

- Hooks may only output `provides` entries to the build plan. They must not output `requires`.
- Hooks must all proceed regular buildpacks in `order` definitions (e.g., in `builder.toml`).
- Hooks are always `optional`.

Hooks generate Dockerfiles before the regular buildpack build phase.
To generate these Dockerfiles, the lifecycle executes the hook's `/bin/build` executable with the same interface as regular buildpacks.
However,
- Hooks `/bin/build` must not write to the app directory.

> - Hooks `/bin/build` may be executed in parallel.

Reviewer 9: Can we instead call it `extend` or something?

Author: https://github.com/buildpacks/rfcs/pull/173#discussion_r679525185

Reviewer 5: This is now `/bin/generate`. I'm going to mark this as resolved.

Finally, base images may be statically labeled with any number of `provides` that are treated as build plan entries.
These `provides` may contain fields other than `name`, which, when mismatched with `requires`, mark the entry as `unmet` by the stack.
This is important to ensure that:
- Rebasing always remains an option for end users.
- Buildpacks do not become dependent on hooks.
- Builds can be time-optimized by creating base images ahead of time.

NOTE: the above can be accomplished by a hook with a no-op `/bin/build` -- do we really need this?

#### Example: App-specified Dockerfile Hook

This example hook would allow an app to provide runtime and build-time base image extensions as "run.Dockerfile" and "build.Dockerfile."
The app developer can decide whether the extensions are rebasable.

##### `/cnb/hooks/com.example.apphook/bin/build`
```
#!/bin/sh

> [ -f build.Dockerfile ] && cp build.Dockerfile "$1/"

Reviewer 9: Should we keep this as an env var instead of `$1`?

env vars are typically more flexible as we can make more changes in the future?

Author: @ekcasey's suggestion was to keep the hook interface as similar to the buildpack interface as possible, and then roll out additional changes (like moving positional args to env vars) to both at the same time.

Reviewer 5: This seems resolved, I am going to hide it - but @samj1912 please re-open if you would like to discuss further.

[ -f run.Dockerfile ] && cp run.Dockerfile "$1/"
```

#### Example: RPM Dockerfile Hook (app-based)

This example hook would allow a builder to install RPMs for each language runtime, based on the app directory.

Note: The Dockerfiles referenced must disable rebasing, and build times will be slower compared to buildpack-provided runtimes.

##### `/cnb/hooks/com.example.rpmhook/bin/build`
```
#!/bin/sh
[ -f Gemfile.lock ] && cp "$BUILDPACK_DIR/Dockerfile-ruby" "$1/Dockerfile"
[ -f package.json ] && cp "$BUILDPACK_DIR/Dockerfile-node" "$1/Dockerfile"
```

### Dockerfiles for Base Images

The same Dockerfile format may be used to create new base images or modify existing base images outside of the app build process (e.g., before creating a builder). Any specified labels override existing values.

> Dockerfiles that are used to create a base image must create a `/cnb/image/genpkgs` executable that outputs a [CycloneDX](https://cyclonedx.org)-formatted list of packages in the image with PURL IDs when invoked. This executable is executed after all Dockerfiles are applied, and the output replaces the label `io.buildpacks.sbom`. This label doubles as a Software Bill-of-Materials for the base image. In the future, this label will serve as a starting point for the application SBoM.

Reviewer 9: Can we just use the `build`/`extend` script to output this instead?

Author: The original SBoM needs to be replaced after additional OS packages are installed, so both mechanisms are needed.

(also answered during WG)

Reviewer 5: /queue-issue buildpacks/lifecycle "Lifecycle run /cnb/image/genpkgs on the extended run image" epic/dockerfiles type/enhancement

Reviewer 5: Marking as resolved given that we're leaving SBOM generation up to the platform.

- Depends on: [RFC#172](https://github.com/buildpacks/rfcs/pull/172)

# Summary
[summary]: #summary

This RFC introduces functionality for customizing base images, as an alternative to stackpacks ([RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md)).

# Motivation
[motivation]: #motivation

[RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md) introduces complexity by defining an API that allows buildpacks to modify base images. To avoid this complexity, we could rely on generated Dockerfiles for base image manipulation. This would simplify the original proposal by, e.g., only requiring a copy of the buildpack on the build-time base image.

# What it is
[what-it-is]: #what-it-is

This RFC proposes that we replace stackpacks with dynamically-generated build-time and runtime Dockerfiles that act as pre-build hooks.
These hooks would participate in detection and execute before the buildpack build process.

For a given application, a build that uses hooks could be optimized by creating a more narrowly-scoped builder that does not contain hooks.

# How it Works
[how-it-works]: #how-it-works

> Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 

Reviewer 10: And what about Podman, is it counted?

Reviewer 12: it works reasonably well for me on Linxu, however the VM podman in macOS and Windows not so much

Reviewer 11: IIUC, Podman uses buildah under the hood for building containers and provides only a subset of buildah's functionalities so if the buildah is mentioned, we should be fine.

Author: Intent is not to be restrictive. This is just clarifying that existing build tools that support the Dockerfile format may be used to implement this RFC.

Reviewer 6: is this meant to say a platform like VMWare/Heroku/Google would _provide_ the thing that applies Dockerfiles..or is this saying that `lifecycle` is able to choose any of these tools?

Reviewer 5: ```suggestion
Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. The order of operations would be something like the following:
* analyze
* detect
* restore
* <new lifecycle phase> run hooks' bin/build, output Dockerfiles are written to a volume
* <new lifecycle phase OR platform> apply Dockerfiles to run image (could run in parallel with below two)
* <new lifecycle phase OR platform> apply Dockerfiles to build image
* build
* export
```

Reviewer 5: Now that I'm thinking about it, we could run bin/build for hooks during `detect`, since hooks don't use the cache...

### Dynamically-applied Dockerfiles

> A builder image may include any number of "hook" directories in `/cnb/hooks/`.

Reviewer 4: Rather than introducing the "hook" concept, could we tie this to any existing concept (especially if it's something in Docker). For example, the [`FROM` instruction](https://docs.docker.com/engine/reference/builder/#from) initializes a new build stage and sets the Base Image for subsequent instructions. 

Can we tie this concept to "stage", or "base-image" or even "from"?

The other problem with "hook" is that it implies (to me) that it's some asynchronous fire-and-forget thing like a Webhook.

Reviewer 4: After discussion in the Working Group, I would be comfortable with any of the following alternatives:
* "ext" - `/cnb/ext/`, `ext.toml`
* "from" - `/cnb/from/`, `from.toml`
* "stage" - `/cnb/stages/`, `stage.toml`

I would prefer `stage`, although I wish there was some way to make it clear that these were stages _before_ the buildpack build.

Reviewer 8: +1 on `staging`. I'm also in favor of something like `ext` as proposed by @jabrown85 and @samj1912 .

I'm tying this in with the idea of providing some differentiator (suffix) for files shared with the existing buildpack API.

Here's an example of what the dir structure may look like for either cases.

### `staging` with suffixed file extensions

##### Pros

* Clear_ish_ about its purpose to "stage" the environment.

##### Cons

* Limiting to being only a "pre" build process. May we want "post" processing in the future?

##### Prior Art

* Loose tie to "stages" in Dockerfiles (see `FROM` and multi-stage).

```text
.
└── cnb/
    ├── staging/
    │   ├── stage-w-dir-only/
    │   │   └── 0.1.0/
    │   │       ├── launch.stage.toml
    │   │       ├── build.stage.toml
    │   │       ├── Dockerfile
    │   │       └── staging.toml
    │   └── stage-w-dynamic-dockerfile/
    │       └── 1.2.3/
    │           ├── bin/
    │           │   ├── build
    │           │   └── detect
    │           └── staging.toml
    └── buildpacks/
        └── buildpack-1/
            └── 0.1.0/
                ├── bin/
                │   ├── build
                │   └── detect
                └── buildpack.toml
```

### `ext` with suffixed file extensions

##### Pros

* Flexible to provide additional future functionality.

##### Cons

* A little ambiguous as `ext`.
* If we use `extensions`, then it may collide in terms to "[extension](https://github.com/buildpacks/spec/tree/main/extensions)" specifications.

##### Prior Art

* Kernel extensions ([kext](https://support.apple.com/guide/deployment-reference-macos/kernel-extensions-in-macos-apd37565d329/web))

```text
.
└── cnb/
    ├── ext/
    │   ├── ext-w-dir-only/
    │   │   └── 0.1.0/
    │   │       ├── launch.ext.toml
    │   │       ├── build.ext.toml
    │   │       ├── Dockerfile
    │   │       └── ext.toml
    │   └── ext-w-dynamic-dockerfile/
    │       └── 1.2.3/
    │           ├── bin/
    │           │   ├── build
    │           │   └── detect
    │           └── ext.toml
    └── buildpacks/
        └── buildpack-1/
            └── 0.1.0/
                ├── bin/
                │   ├── build
                │   └── detect
                └── buildpack.toml
```

Reviewer 5: I like /cnb/stages/, stage.toml. I'm sorry @jromero but when there are so many `.`s in the filename I can no longer pay attention 🙃

Reviewer 8: Based on today's WG, we have a few names to vote on from this [list](https://docs.google.com/document/d/1sq5TpJO-m8aTzgbAiVlS1QdNovne-3ku9FxxC2S_Zs4/edit#heading=h.4sikugimndzz).

Using emojis vote for your favorite:

🎉  = extensions
❤️ = stages
🚀 = initializers

Reviewer 13: initializers is clear about the "when" (extensions and stages are ambiguous in this sense), and it also leaves room for hooks at the other parts of the lifecycle (e.g. finalizers)

Author: To me, "initializer" implies that the changes are scoped to initializing the build process. The changes applied the run image don't initialize anything.

> * restore
> * <new lifecycle phase> run extensions' bin/build, output Dockerfiles are written to a volume

Reviewer 5: ```suggestion
* <new lifecycle phase> run extensions' bin/build, output Dockerfiles are written to a volume
* restore

```

Hooks are similar to buildpacks: they have a `/bin/build` and `/bin/detect` executable.
However, instead of a `buildpack.toml` file, hooks have a `hook.toml` file:
```toml

> api = "<buildpack API version>"

Reviewer 4: Would it make sense to have some kind of version-variant here like `api = "0.8-hook"`?

Author: They are different files, but the spec could enforce the similarity.

### Dynamically-applied Dockerfiles

> A builder image may include any number of "extensions" directories in `/cnb/ext/`.

Reviewer 5: ```suggestion
A builder image may include any number of "extensions" directories in `/cnb/extensions/`.
```

name = "<hook name>"
version = "<hook version>"
homepage = "<hook homepage>"
description = "<hook description>"
keywords = [ "<string>" ]

[[hook.licenses]]
type = "<string>"
uri = "<uri>"
```

Hooks may be packaged and examined similar to buildpacks, but with analogous `pack hook` subcommands.

Other pack CLI commands, such as `pack builder create`, will be extended to include support for hooks.

Unlike buildpacks,
- Hooks must not be included in a meta-buildpacks
- Hooks must not have `order`/`group` definitions in `hook.toml`

> Hooks participate in the buildpack detection process, with the same interface for `/bin/detect`.

Reviewer 5: /queue-issue buildpacks/lifecycle "Detector should run hooks" epic/dockerfiles type/enhancement

However,

> - `/bin/detect` is optional for hooks, and they are assumed to pass detection when it is not presetn.

Reviewer 8: ```suggestion
- `/bin/detect` is optional for hooks, and they are assumed to pass detection when it is not present.
```

> - `/bin/detect` is optional for hooks, and they are assumed to pass detection when it is not present. Just like with buildpacks, a /bin/detect that exits with a 0 exit code passes detection, and fails otherwise.

Reviewer 6: Is it documented what user `bin/detect` is executed at for hooks?

Author: Added -- UID/GID is the same as regular buildpacks for both build and detect

- Hooks may only output `provides` entries to the build plan. They must not output `requires`.

> - Hooks must all proceed regular buildpacks in `order` definitions (e.g., in `builder.toml`).

Reviewer 8: This might be my lack of understanding of the English language but what does "proceed" here mean? Does there need to be more clarification?

Reviewer 6: I think this is meant to be `precede`

Reviewer 6: Can we get a `builder.toml` example somewhere in the rfc? Is there a new `[hooks]` table?

Author: A separate `[[extensions]]` table seems ideal to me, given that all extensions must come first.

This RFC isn't intended to be exhaustively cover the pack CLI UX, but if you have strong opinions about `builder.toml`, could you propose a suggestion?

- Extensions must not be included in a meta-buildpacks

> - Extensions must not have `order`/`group` definitions in `extension.toml`

Reviewer 5: @sclevine could you elaborate on the reason for this restriction?

Reviewer 4: The order/group would make it a meta-buildpack. I believe the restriction on meta-buildpack is there to reduce complexity.

Hooks generate Dockerfiles before the regular buildpack build phase.
To generate these Dockerfiles, the lifecycle executes the hook's `/bin/build` executable with the same interface as regular buildpacks.
However,

> - Hooks `/bin/build` must not write to the app directory.

Reviewer 4: Can we enforce this (i.e. by not mounting the app dir)?

Author: Do you mean mounting it as read-only?

Author: We currently don't enforce read-only behavior with pack elsewhere (e.g., for /bin/detect), but I agree that we should.

- Hooks `/bin/build` may be executed in parallel.

> - Hooks `<layers>` directory is replaced by an `<output>` directory.

Reviewer 4: Can we also set working dir to `<output>` dir? That might be weird (the hook's own root as cwd), but it makes the absence of app dir less weird.

Author: App dir is present (read-only)

> - `/bin/detect` is optional for extensions, and they are assumed to pass detection when it is not present. Just like with buildpacks, a /bin/detect that exits with a 0 exit code passes detection, and fails otherwise.

Reviewer 5: The idea of a static build plan seems to make sense in light of what we're doing for the `generate` step.

> - Extensions may only output `provides` entries to the build plan. They must not output `requires`.

Reviewer 5: @sclevine could you elaborate on the reason for this restriction?

Reviewer 4: I believe this avoids having an extension with dependencies (and avoids cases where you might want a regular buildpack to provider a required dep).

> - Extensions must all precede regular buildpacks in `order` definitions (e.g., in `builder.toml`).

Reviewer 5: @sclevine on digging into this a bit further, I'm confused as to how this would work. As I understand, extensions are not allowed to output `requires` to the build plan - they can only output `provides`. However in resolving the build plan, the lifecycle currently requires all `requires` to be [provided first](https://github.com/buildpacks/lifecycle/blob/b6408639c0b0daada589b7b59a8eda2631bc4738/detector_test.go#L856). This would imply that extensions should come **last** in the order definition.

Separately in [this comment](https://github.com/buildpacks/samples/pull/113#discussion_r726206827) it was suggested to have a separate `[[order.ext]]` table in builder.toml. How would this work, if the detector is iterating through each group in the order and choosing the first one where all non-optional buildpacks pass detect? If there are two orders, that implies two detects - in which case, extensions and buildpacks cannot collaborate on the build plan.

Reviewer 5: From an implementation perspective, the simplest thing would be to allow extensions to appear anywhere in the order, but add an `extension = <bool>` key to [[[order.group]]](https://github.com/buildpacks/spec/blob/74f91666fc22e7568acdd0ca9e2e314ff150d740/platform.md#ordertoml-toml).

Reviewer: @natalieparellano 

> However in resolving the build plan, the lifecycle currently requires all requires to be provided first.

Despite the confusing title on that test providing buildpacks need to precede requiring buildpacks in the order. For example a maven buildpacks requires a JDK to execute, so the buildpack that provides the JDK must precede the maven buildpack in the order.

> From an implementation perspective, the simplest thing would be to allow extensions to appear anywhere in the order, but add an extension = <bool> key to [[order.group]].

Detect order should match execution order. If we interleave extensions and regular buildpacks we won't be able to run the extensions in a separate phase that generates the build image prior to the build phase.

Reviewer 5: We also discussed with @sclevine how the `[[order-ext]]` table could be prepended to the detect order as a meta buildpack, allowing buildpacks and extensions to collaborate on the build plan while keeping the tables separate.

Reviewer 5: Marking this as resolved and added a code suggestion to make it clearer.

> - Extensions must all precede regular buildpacks in `order` definitions (e.g., in `builder.toml`).

Reviewer 5: ```suggestion
- Extensions are not included in `order` definitions (e.g., in `builder.toml`); instead, a separate `order-ext` table should be used.
```

Reviewer 5: Do we also need a `group-ext`?

- Extensions are always `optional`.

Extensions generate Dockerfiles before the regular buildpack build phase.

> To generate these Dockerfiles, the lifecycle executes the extension's `/bin/build` executable with the same UID, GID, and interface as regular buildpacks.

Reviewer: stale comment, came along for the ride by accident. apologies.

> To generate these Dockerfiles, the lifecycle executes the extension's `/bin/build` executable with the same UID, GID, and interface as regular buildpacks.

Reviewer 5: ```suggestion
To generate these Dockerfiles, the lifecycle executes the extension's `/bin/generate` executable with the same UID, GID, and interface as regular buildpacks.
```

However,

> - Extensions `/bin/build` must not write to the app directory.

Reviewer 5: ```suggestion
- Extensions `/bin/generate` must not write to the app directory.
```

> - Extensions `/bin/build` may be executed in parallel.

Reviewer 5: ```suggestion
```
Now that we are using the build plan, this is no longer possible.

- Extensions `<layers>` directory is replaced by an `<output>` directory.

> - If an extension is missing `/bin/build`, the extension root is treated as a pre-populated `<output>` directory.

Reviewer 5: ```suggestion
- If an extension is missing `/bin/generate`, the extension root is treated as a pre-populated `<output>` directory.
```

- Either `Dockerfile` or either or both of `build.Dockerfile` and `run.Dockerfile`

Support for other instruction formats, e.g., LLB JSON files, could be added in the future.

`build.Dockerfile`, `run.Dockerfile`, and `Dockerfile` target the builder image, runtime base image, or both base images, respectively.

If no Dockerfiles are present, `/bin/build` may still consume build plan entries and add metadata to `build.toml`/`launch.toml`.

> Dockerfiles are applied to their corresponding base images after all hooks are executed and before any regular buildpacks are executed.

Reviewer 5: /queue-issue buildpacks/lifecycle "Lifecycle apply Dockerfiles to build image" epic/dockerfiles type/enhancement
/queue-issue buildpacks/lifecycle "Lifecycle apply Dockerfiles to run image" epic/dockerfiles type/enhancement

Dockerfiles are applied in the order determined during buildpack detection.

> All Dockerfiles are provided with `base_image` and `build_id` args.

Reviewer 8: Could you provide a more concrete use case for how `build_id` could be used? Who is able to provide this value to invalidate the cache?

Author: The lifecycle provides it automatically.
Example is here: https://github.com/buildpacks/rfcs/pull/173/files#diff-aeaa4e3f3765f6b7ec295a8c92881bdeae1313f69fd338da209548883dafdf00R180

> - Either `Dockerfile` or either or both of `build.Dockerfile` and `run.Dockerfile`

Reviewer 5: Per 7/7/22 Working Group, removed prefix-less `Dockerfile` in 3a26570

The `base_image` arg allows the Dockerfile to reference the original base image.

> The `build_id` arg allows the Dockerfile to invalidate the cache after a certain layer and must be defaulted to `0`.

Reviewer 5: How are we keeping track of what this value should be? Does it go in a label on the app image?

Author: Maybe this should be a UUID explicitly so that it doesn't need to be tracked?

Reviewer 5: ```suggestion
The `build_id` arg allows the Dockerfile to invalidate the cache after a certain layer and must be defaulted to `0`. The executor of the Dockerfile will provide the `build_id` as a UUID (this eliminates the need to track this variable). 
```

> `build.Dockerfile`, `run.Dockerfile`, and `Dockerfile` target the builder image, runtime base image, or both base images, respectively.

Reviewer 5: ```suggestion
`build.Dockerfile`, `run.Dockerfile`, and `Dockerfile` target the builder image, runtime base image, or both base images, respectively. Dockerfiles that target the build image must begin with the following lines: `ARG base_image\nFROM ${base_image}`
```

Reviewer 6: Hmm - how would this work with multi-stage Dockerfiles? 

```
FROM some-registry/repo/image as external_dep

FROM ${base_image}
COPY --from=external_dep ...
```

Reviewer 6: I think we need to disallow multi-stage and essentially ignore `FROM` clauses and that we execute the commands directly on the running build container.

Reviewer 5: Hmm I don't follow - the example you gave doesn't start with 
```
ARG base_image
FROM ${base_image}
```
so I think it would be disallowed.

Reviewer 6: I'm thinking about multi-stage builds and maybe we should explicitly disallow them for `build` exts or else we need to parse the `Dockerfile` and understand the last stage to be `base_image`, right?

This would pass the validation described, but still ultimately ends up trying to use a different build base image.

```
ARG base_image
FROM ${base_image}
# do nothing to pass validation

FROM some-registry/repo/image as external_dep

FROM ${base_image}
COPY --from=external_dep ...
RUN some-command

FROM alpine
RUN echo hmm
```

> When the `$build_id` arg is referenced in a `RUN` instruction, all subsequent layerrs will be rebuilt on the next build (as the value will change).

Reviewer 8: ```suggestion
When the `$build_id` arg is referenced in a `RUN` instruction, all subsequent layers will be rebuilt on the next build (as the value will change).
```

> If no Dockerfiles are present, `/bin/build` may still consume build plan entries and add metadata to `build.toml`/`launch.toml`.

Reviewer 5: Updated in f8e7bfa

Dockerfiles are applied to their corresponding base images after all extensions are executed and before any regular buildpacks are executed.

> Dockerfiles are applied in the order determined during buildpack detection.

Reviewer 5: I have some questions about how multiple Dockerfiles are supposed to be chained together. 

Let's say after running extensions we have two build.Dockerfile's in this order:

```
ARG base_image
FROM ${base_image}

RUN echo "hello" > world.txt
```
```
ARG base_image
FROM ${base_image}

RUN cat world.txt > again.txt
```
What is the expected behavior when applying the second build.Dockerfile - should it succeed? If so then the `ARG` it receives is not the "original" build image but the intermediate image after applying the first build.Dockerfile, is that correct?

Assuming that's correct, and thinking about extending the build image "in container" with kaniko - we'd need to parse every build.Dockerfile that we receive and chain all the instructions together as if it were one big Dockerfile. There are some additional complexities with respect to multi-stage builds, but we haven't quite gotten to the point of thinking about those yet. It would be helpful though to get confirmation if we're on the right track.

Now let's say after running extensions we have two build.Dockerfile's in this order:

```
ARG base_image
FROM ${base_image}

RUN echo "hello" > world.txt
```
```
FROM alpine

RUN echo "hello" >> world.txt
```

How many "hello"s should exist in world.txt after both build.Dockerfiles are applied - just one? I'm assuming if we see a `FROM some-image` in any Dockerfile (not using the provided `base_image` arg), then all the Dockerfiles that come before it have no effect and should be ignored. Is that correct?

Reviewer 6: Here are my thoughts...

For a `build` extension, the `base_image` would indeed be equal to "The current build image state" as the commands would be executed one after the other, in order of buildpack exts defined in `order.toml`. If we expect a clean slate after each `Dockerfile`, that would mean we need to run each `Dockerfile` in their own container and carry forward the snapshots to be re-assembled. That seems like it would be pretty slow and not worth the complexity IMO.

Your second example, I would think for a `build` extension you _must_ use `base_image` OR we need to always ignore the `FROM` for `Dockerfile.build` extensions and _just_ run the remaining instructions on the current state of the build container. The `Dockerfile.build` would essentially just be the transport for commands to run, The RFC has examples of using `Dockerfile` to create a new runtime base image, but I don't think we said you could change the build image base anywhere in the RFC.  The more I think about `build` extensions, the more I think we should _require_ that they are `FROM scratch`. Otherwise aren't we having kaniko executor pull/extract the base build layers as it processes N number of `Dockerfile.build` anyway? 

Now, as far as `Dockerfile.run` goes - this doesn't really work. If we intend on letting `Dockerfile.run` choose a `base_image` as described in the RFC...doesn't that mean we can really only support a single `Dockerfile.run` extension?

Imagine you have 3 `Dockerfile.run`s. One that calls `apt-get install x`, one that adds a file to `/opt` and one that changes the `base_image` to `scratch` from the previous target of `ubuntu:bionic`. If we executed them in that order, the `apt` installation would obviously not be valid as it would essentially get transplanted onto a scratch image. I guess we could have some restrictions around  `base_image` - saying it must be the same distro/version as the builder's target run-image...but that means `scratch` is off the table I think.

Similar to the first concern you had, I had _assumed_ all `Dockerfile.run` would be executed in the same container one after the other _on the target run image_. That doesn't work, however, if they all use different `base_image`. One, we don't know until execution what image we should be running on. We would need to run `kaniko/executor` and have it extract the `base_image` into the running container. We would also need to clean and extract the _next_ `base_image` to keep them isolated. Even then, we would end up making a resulting base image that would likely not function if they all used different `base_image`. They would also potentially have to duplicate work (`apt-get update`). I think sharing a container makes sense, but it means we need to know the base image ahead of time I think.

Reviewer 5: Added https://github.com/buildpacks/rfcs/pull/173/files#r793096512 for clarity.

> Dockerfiles are applied in the order determined during buildpack detection.

Reviewer 5: ```suggestion
Dockerfiles are applied in the order determined during buildpack detection. When multiple Dockerfiles are applied, the intermediate image generated from the application of the current Dockerfile will be provided as the `base_image` ARG to the next Dockerfile. Dockerfiles that target the run image (only) may ignore the provided `base_image` (e.g., `FROM some-other-image`). 
```

> A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before a runtime Dockerfile is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.

Reviewer 5: If there are multiple `run.Dockerfile`, we should verify that ALL of them set this label to `true` before allowing the app image to have `true`.

Reviewer 5: > `io.buildpacks.rebasable=false` is set automatically
by whom?

Author: By whatever applies the Dockerfile to the base image. Intent is that `LABEL io.buildpacks.rebasable=true` must be present to set the value to true, even if `io.buildpacks.rebasable=true` is already set on the runtime base image.

Reviewer 5: ```suggestion
A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically on the base image before a runtime Dockerfile is applied and must be explicitly set to `true` if desired. If multiple Dockerfiles are applied, all must set `io.buildpacks.rebasable=true` for the final value to be `true`. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
```

Reviewer 5: On second thought, what about having `rebasable = true` be part of launch.toml? That might make it easier to iterate through the hooks and determine if all of them set it to true.

> Finally, base images may be statically labeled with any number of `provides` that are treated as build plan entries.

Reviewer 8: When we describe "statically labeled" are we talking about a label on the image or the `provides` that is generated by `/bin/detect`?

Reviewer 5: Does this apply to both the build and run images?

Reviewer 5: From WG: yes. However there is no concept (from the lifecycle's perspective) of `build:` and `run:` (like with mixins). Buildpacks may use `[requires.metadata]` to communicate to hooks where exactly they need a certain dependency.

Author: Removed

> These `provides` may contain fields other than `name`, which, when mismatched with `requires`, mark the entry as `unmet` by the stack.

Reviewer 8: What `requires` are we referring to here? Based on earlier text hooks can't declare `requires`. So are these buildpack `requires`?

Reviewer 5: > So are these buildpack `requires`

That was my reading of it.

Reviewer 2: I had to re-read this paragraph a few times and I'm still not totally sure I understand what's trying to be conveyed here. Maybe it could be clarified a bit or an example added?

Author: Removed

> The `build_id` arg allows the Dockerfile to invalidate the cache after a certain layer and must be defaulted to `0`. The executor of the Dockerfile will provide the `build_id` as a UUID (this eliminates the need to track this variable). 

Reviewer 5: ```suggestion
The `build_id` arg allows the Dockerfile to invalidate the cache after a certain layer. The executor of the Dockerfile will provide the `build_id` as a UUID (this eliminates the need to track this variable). 
```

Reviewer 5: A default doesn't make sense if we're using a UUID

Author: The default is just so that nothing blows up if whatever builds the Dockerfile doesn't provide the value.

Build args specified in `build.toml` are provided to `build.Dockerfile` or `Dockerfile` (when applied to the build-time base image).
Build args specified in `launch.toml` are provided to `run.Dockerfile` or `Dockerfile` (when applied to the runtime base image).

> A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically on the base image before a runtime Dockerfile is applied and must be explicitly set to `true` if desired. If multiple Dockerfiles are applied, all must set `io.buildpacks.rebasable=true` for the final value to be `true`. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.

Reviewer 5: As we're working through the [POC](https://github.com/buildpacks/lifecycle/issues/709) it seems advantageous to keep the "extender" binary/phase pretty ignorant of CNB business logic i.e., it would be great if the extender could just apply the Dockerfiles (and corresponding build args) it is provided, without having to worry about checking the value of this label after each application. Maybe there could be a 
```
[labels]
rebasable = <bool>
```
table in build.toml and launch.toml? This would save the lifecycle (non-extender binary) the trouble of having to parse any Dockerfiles. The downside is that it would be harder for application developers to specify this value.

Reviewer 5: After talking this over we think the `LABEL` instruction is the best way to communicate "rebasable" because it is closest to the logic (the Dockerfile instructions) that determine whether rebasable should be true or false.

NOTE: the above can be accomplished by a hook with a no-op `/bin/build` -- do we really need this?

#### Example: App-specified Dockerfile Hook

This example hook would allow an app to provide runtime and build-time base image extensions as "run.Dockerfile" and "build.Dockerfile."
The app developer can decide whether the extensions are rebasable.

##### `/cnb/hooks/com.example.apphook/bin/build`
```
#!/bin/sh
[ -f build.Dockerfile ] && cp build.Dockerfile "$1/"
[ -f run.Dockerfile ] && cp run.Dockerfile "$1/"
```

#### Example: RPM Dockerfile Hook (app-based)

This example hook would allow a builder to install RPMs for each language runtime, based on the app directory.

Note: The Dockerfiles referenced must disable rebasing, and build times will be slower compared to buildpack-provided runtimes.

##### `/cnb/hooks/com.example.rpmhook/bin/build`
```
#!/bin/sh

> [ -f Gemfile.lock ] && cp "$BUILDPACK_DIR/Dockerfile-ruby" "$1/Dockerfile"
> [ -f package.json ] && cp "$BUILDPACK_DIR/Dockerfile-node" "$1/Dockerfile"

Reviewer 5: ```suggestion
[ -f Gemfile.lock ] && cp "$CNB_BUILDPACK_DIR/Dockerfile-ruby" "$1/Dockerfile"
[ -f package.json ] && cp "$CNB_BUILDPACK_DIR/Dockerfile-node" "$1/Dockerfile"
```
(I think this is what you meant)

> ##### `/cnb/ext/com.example.appext/bin/build`

Reviewer 5: ```suggestion
##### `/cnb/ext/com.example.appext/bin/generate`
```

> ##### `/cnb/ext/com.example.rpmext/bin/build`

Reviewer 5: ```suggestion
##### `/cnb/ext/com.example.rpmext/bin/generate`
```

### Dockerfiles for Base Images

The same Dockerfile format may be used to create new base images or modify existing base images outside of the app build process (e.g., before creating a builder). Any specified labels override existing values.

> Dockerfiles that are used to create a base image must create a `/cnb/image/genpkgs` executable that outputs a [CycloneDX](https://cyclonedx.org)-formatted list of packages in the image with PURL IDs when invoked. This executable is executed after all Dockerfiles are applied, and the output replaces the label `io.buildpacks.sbom`. This label doubles as a Software Bill-of-Materials for the base image. In the future, this label will serve as a starting point for the application SBoM.

Reviewer 3: A discussion and RFC/WIP has also been created on opencontainer to support SBOM part of an image METADATA using a blob store json object - see: https://github.com/opencontainers/artifacts/pull/37
So instead of putting the SBOM json file generated as `io.buildpacks.sbom` we could rely on a more generic approach relying on image METADATA. WDYT ? @sclevine

Author: I agree that `io.buildpacks.sbom` is not a long term solution. I expected https://github.com/buildpacks/rfcs/pull/166 to specify a more complete solution that included both the stack and buildpack SBoM. It looks like the base image SBoM was de-scoped from #166, so we need a follow-up RFC to replace `io.buildpacks.sbom` and specify that the base image SBoM should be merged into the final SBoM. 

Until https://github.com/opencontainers/artifacts/pull/37 lands and is widely supported in registries, I think we should use cosign's SBoM support: https://github.com/sigstore/cosign/blob/main/specs/SBOM_SPEC.md

But before we implement either of those standardized solutions, maybe we should move the stack SBoM to an FS layer and ensure that it's merged into SBoM generated in #166?

Reviewer 3: > Until [opencontainers/artifacts#37](https://github.com/opencontainers/artifacts/pull/37) lands and is widely supported in registries, I think we should use cosign's SBoM support: https://github.com/sigstore/cosign/blob/main/specs/SBOM_SPEC.md

That will request additional work then to migrate from cosign to opencontainer spec when we will move and to implement a mechanism able to convert cosign SBOM to opencontainer SBOM ...

Author: > That will request additional work then to migrate from cosign to opencontainer spec when we will move and to implement a mechanism able to convert cosign SBOM to opencontainer SBOM ...

What is the alternative? I don't think we should implement the OCI version until it is merged and implemented in popular registries, which I don't expect to happen in the near future.

(And as @dlorenc mentions, it's just the storage mechanism, not the format of the SBoM itself.)

Reviewer 5: Can we clarify the interplay between `genpkgs` and any sbom files output by hooks? It's kind of confusing that both mechanisms are allowed. When would you use one vs the other?

Reviewer 5: Marking as resolved given that we're leaving SBOM generation up to the platform (see https://github.com/buildpacks/spec/pull/298#issuecomment-1099543096).

> Dockerfiles that are used to create a base image must create a `/cnb/image/genpkgs` executable that outputs a [CycloneDX](https://cyclonedx.org)-formatted list of packages in the image with PURL IDs when invoked. This executable is executed after all Dockerfiles are applied, and the output replaces the label `io.buildpacks.sbom`. This label doubles as a Software Bill-of-Materials for the base image. In the future, this label will serve as a starting point for the application SBoM.

Reviewer 5: ```suggestion
Base images may contain a layer with SBoM files as detailed in https://github.com/buildpacks/rfcs/pull/186. After Dockerfiles are applied, the final step in the extension process will be to run a `genpkgs` executable (supplied by the platform, see further details that follow) in the context of the extended image that will generate an updated SBoM, taking as input any files that were already present. The diffID of the layer containing the output from `genkpgs` will replace the `io.buildpacks.base.sbom` label on the extended run image. 

When assembling layers for the extended run image, the extender will exclude the `genpkgs` executable path, so that `genpkgs` will not be present in the extended run image.

The project will ship a reference implementation of `genpkgs` in the [lifecycle image](https://hub.docker.com/r/buildpacksio/lifecycle), a minimal image containing the lifecycle binaries. Platforms when running the extender may choose to override the provided `genpkgs` with their own implementation. We propose that the reference implementation use [syft](https://github.com/anchore/syft).
```

Reviewer 5: @samj1912 @sclevine updated with my understanding from yesterday's OH...

Reviewer 5: > Can we clarify the interplay between genpkgs and any sbom files output by hooks? It's kind of confusing that both mechanisms are allowed. When would you use one vs the other?

I'm still curious about this...

### Example Dockerfiles

Dockerfile used to create a runtime base image:

```
ARG base_image
FROM ${base_image}
ARG build_id=0

LABEL io.buildpacks.image.distro=ubuntu
LABEL io.buildpacks.image.version=18.04
LABEL io.buildpacks.rebasable=true

ENV CNB_USER_ID=1234
ENV CNB_GROUP_ID=1235

RUN groupadd cnb --gid ${CNB_GROUP_ID} && \
  useradd --uid ${CNB_USER_ID} --gid ${CNB_GROUP_ID} -m -s /bin/bash cnb

USER ${CNB_USER_ID}:${CNB_GROUP_ID}

> 
> COPY genpkgs /cnb/image/genpkgs

Reviewer 5: ```suggestion
```

ARG build_id=0

RUN echo ${build_id}

RUN apt-get update && apt-get install -y curl && rm -rf /var/lib/apt/lists/*
```
(note: this Dockerfile disables rebasing, as OS package installation is not rebasable)

`run.Dockerfile` for use with the example `app.run.Dockerfile.out` extension that installs a special package to /opt:
```
ARG base_image
FROM ${base_image}
ARG build_id=0

LABEL io.buildpacks.rebasable=true

RUN curl -L https://example.com/mypkg-install | sh # installs to /opt
```
(note: rebasing is explicitly allowed because only a single directory in /opt is created)

# Drawbacks
[drawbacks]: #drawbacks

- Involves breaking changes.

# Alternatives
[alternatives]: #alternatives

- Stackpacks
- Previous versions of this RFC that don't interact with the buildpack API.

# Unresolved Questions
[unresolved-questions]: #unresolved-questions

- Should `genpkgs` be part of this proposal? Opinion: Yes, otherwise it's difficult to maintain a valid SBoM.
- Should we allow base images to provide build plan entries so that extensions aren't required to satisfy buildpacks? Opinion: Not yet, no-op extension can be used for now.

# Spec. Changes (OPTIONAL)
[spec-changes]: #spec-changes

> This RFC requires extensive changes to all specifications.

Reviewer 14: ```suggestion
This RFC requires extensive changes to all specifications. [buildpacks/spec#267](https://github.com/buildpacks/spec/pull/267).
```

Reviewer 5: Marking as resolved as the linked PR is now closed. I made an alternative suggestion.

These extensions would participate in detection and execute before the buildpack build process.

For a given application, a build that uses extensions could be optimized by creating a more narrowly-scoped builder that does not contain extensions.

# How it Works
[how-it-works]: #how-it-works

Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. The order of operations would be something like the following:
* `analyze`
* `detect` - after standard detection detect will also run extensions' bin/generate, output Dockerfiles are written to a volume
* `extend` - applies run.Dockerfiles to run image (could run in parallel with builder image extension)
* `extend` - applies build.Dockerfiles to builder image
* `restore`
* `build`
* `export`

When Dockerfiles are used to update the run image, care should be taken to ensure registry access prior to the `build` phase, to avoid long builds that fail due to incorrect credentials.

### Dynamically-applied Dockerfiles

A builder image may include any number of "extensions" directories in `/cnb/extensions/`.

> Extensions are similar to buildpacks: they have two executables: `/bin/detect` and `/bin/generate`. The interface for these executables is similar to a buildpack's `/bin/detect` and `/bin/build`.

Reviewer: ```suggestion
Extensions are similar to buildpacks: they have two executables: `./bin/detect` and `./bin/generate`. The interface for these executables is similar to a buildpack's `./bin/detect` and `./bin/build`.
```

Starting with a `/` always bugged me a bit because it looks like an absolute path when it should be relative to the extension root. Making the directory paths explicitly relative (e.g. `./detect/`) makes this more confusing imho.

However, instead of a `buildpack.toml` file, extensions have a `extension.toml` file:
```toml
api = "<buildpack API version>"

[extension]
id = "<extension ID>"
name = "<extension name>"
version = "<extension version>"
homepage = "<extension homepage>"
description = "<extension description>"
keywords = [ "<string>" ]

[[extension.licenses]]
type = "<string>"
uri = "<uri>"
```

Extensions may be packaged and examined similar to buildpacks, but with analogous `pack extension` subcommands.

Other pack CLI commands, such as `pack builder create`, will be extended to include support for extensions.

Unlike buildpacks,
- Extensions must not be included in a meta-buildpacks
- Extensions must not have `order`/`group` definitions in `extension.toml`

> Extensions participate in the buildpack detection process, with the same UID, GID, and interface for `/bin/detect`.
> However,
> - `/bin/detect` is optional for extensions, and they are assumed to pass detection when it is not present. Just like with buildpacks, a /bin/detect that exits with a 0 exit code passes detection, and fails otherwise.
> - If an extension is missing `/bin/detect`, the extension root `./detect` directory is treated as a pre-populated output directory (i.e., extensions can include a static build plan).

Reviewer: ```suggestion
Extensions participate in the buildpack detection process, with the same UID, GID, and interface for `./bin/detect`.
However,
- `./bin/detect` is optional for extensions, and they are assumed to pass detection when it is not present. Just like with buildpacks, a `./bin/detect` that exits with a 0 exit code passes detection, and fails otherwise.
- If an extension is missing `./bin/detect`, the extension root `./detect` directory is treated as a pre-populated output directory (i.e., extensions can include a static build plan).
```

- Extensions may only output `provides` entries to the build plan. They must not output `requires`.
- Extensions are not included in `order` definitions (e.g., in `builder.toml`); instead, a separate `order-extensions` table should be used. The `order-extensions` table will be prepended to each group in the provided `order` (as if `order-extensions` were a composite buildpack).
- Extensions are always `optional`.

Extensions generate Dockerfiles before the regular buildpack build phase.

> To generate these Dockerfiles, the lifecycle executes the extension's `/bin/generate` executable with the same UID, GID, and interface as regular buildpacks.
> However,
> - Extensions `/bin/generate` must not write to the app directory.
> - Extensions `<layers>` directory is replaced by an `<output>` directory.
> - If an extension is missing `/bin/generate`, the extension root `./generate` directory is treated as a pre-populated `<output>` directory.

Reviewer: ```suggestion
To generate these Dockerfiles, the lifecycle executes the extension's `./bin/generate` executable with the same UID, GID, and interface as regular buildpacks.
However,
- Extensions `./bin/generate` must not write to the app directory.
- Extensions `<layers>` directory is replaced by an `<output>` directory.
- If an extension is missing `./bin/generate`, the extension root `./generate` directory is treated as a pre-populated `<output>` directory.
```

> After `/bin/generate` executes, the `<output>` directory may contain

Reviewer: ```suggestion
After `./bin/generate` executes, the `<output>` directory may contain
```

- `build.toml`, with the same contents as a normal buildpack's `build.toml` (the `unmet` table array), but
  - With an additional `args` table array with `name` and `value` fields that are provided as build args to `build.Dockerfile`
- `run.toml`,
  - With an `args` table array with `name` and `value` fields that are provided as build args to `run.Dockerfile`
- Either or both of `build.Dockerfile` and `run.Dockerfile`

Support for other instruction formats, e.g., LLB JSON files, could be added in the future.

`build.Dockerfile` and `run.Dockerfile`target the builder image or runtime base image, respectively.

> If no Dockerfiles are present, `/bin/generate` may still consume build plan entries.

Reviewer: ```suggestion
If no Dockerfiles are present, `./bin/generate` may still consume build plan entries.
```

Hooks are similar to buildpacks: they have a `/bin/build` and `/bin/detect` executable.
However, instead of a `buildpack.toml` file, hooks have a `hook.toml` file:
```toml
api = "<buildpack API version>"

[hook]
id = "<hook ID>"
name = "<hook name>"
version = "<hook version>"
homepage = "<hook homepage>"
description = "<hook description>"
keywords = [ "<string>" ]

[[hook.licenses]]
type = "<string>"
uri = "<uri>"
```

Hooks may be packaged and examined similar to buildpacks, but with analogous `pack hook` subcommands.

Other pack CLI commands, such as `pack builder create`, will be extended to include support for hooks.

Unlike buildpacks,
- Hooks must not be included in a meta-buildpacks
- Hooks must not have `order`/`group` definitions in `hook.toml`

Hooks participate in the buildpack detection process, with the same UID, GID, and interface for `/bin/detect`.
However,
- `/bin/detect` is optional for hooks, and they are assumed to pass detection when it is not present. Just like with buildpacks, a /bin/detect that exits with a 0 exit code passes detection, and fails otherwise.
- Hooks may only output `provides` entries to the build plan. They must not output `requires`.

> - Hooks must all proceed regular buildpacks in `order` definitions (e.g., in `builder.toml`).

Author: ```suggestion
- Hooks must all precede regular buildpacks in `order` definitions (e.g., in `builder.toml`).
```

# Meta
[meta]: #meta
- Name: Support Dockerfiles
- Start Date: 2021-06-30
- Author(s): sclevine
- RFC Pull Request: (leave blank)
- CNB Pull Request: (leave blank)
- CNB Issue: (leave blank)
- Supersedes: [RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md), [RFC#167](https://github.com/buildpacks/rfcs/pull/167)
- Depends on: [RFC#172](https://github.com/buildpacks/rfcs/pull/172)

# Summary
[summary]: #summary

This RFC introduces functionality for customizing base images, as an alternative to stackpacks.

# Motivation
[motivation]: #motivation

Relying on Dockerfiles for base image generation and manipulation allows us to apply buildpacks only to the problem that they solve best: managing application runtimes and dependencies.

# What it is
[what-it-is]: #what-it-is

This RFC proposes that we replace stackpacks with build-time and runtime Dockerfiles that act as pre-build hooks.
These hooks would be equivalent to base image / builder customizations, but would execute immediately before a build.

Most importantly: the changes below do not impact buildpacks or the buildpack API.

> For a given application, a build that uses hooks could be optimized by creating a more narrowly-scoped builder with the hooks pre-applied, without modifying the buildpacks.

Reviewer 3: This is correct. Nevertheless, this RFC should include within the section `What it is` a remark to mention that a dependency exists between what the Hook will do to extend the build image and/or run image and a buildpack.  
example: if the hook build Dockerfile will install a specific tool/lib (JDK, NodeJS, Ruby, Python,...), then the buildpack expects that the hook will extend the build layer of an existing image in order to add the tool/lib which is needed during the execution of the phase `bin/build`

Author: > if the hook build Dockerfile will install a specific tool/lib (JDK, NodeJS, Ruby, Python,...), then the buildpack expects that the hook will extend the build layer of an existing image in order to add the tool/lib which is needed during the execution of the phase bin/build

I don't think a buildpack expects that a *hook* will extend the builder -- just that the builder image has the correct tool/lib available before the build. If the hook is not present, but the builder image has the tool/lib installed, the buildpack would not know the difference.

# How it Works
[how-it-works]: #how-it-works

Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 

### Dynamically-applied Dockerfiles

> A builder image may include any number of "hook" files in `/cnb/hooks.d/`. Hooks may be associated with zero or more buildpack IDs.

Reviewer 3: How a buildpack builder will be able to figure out he different hooks available, how they should be linked to a buildpack,...

**Remark**:  This is why I was mentioning within the `what it is` remark that a dependency exits between a hook and buildpack

Author: > How a buildpack builder will be able to figure out he different hooks available, how they should be linked to a buildpack,...

A buildpack can inspect the base image to determine if hooks have made changes (e.g., could run `python -v` to determine if python is available or needs a `require`)

> Remark: This is why I was mentioning within the what it is remark that a dependency exits between a hook and buildpack

A goal of this RFC is that no explicit dependency exists between hooks and buildpacks. Hooks are just a way of creating narrowly-scoped builder right before the buildpack build process.

Author: After thinking about this more, since the hooks would apply after the detect phase in the most recent proposal, the buildpack wouldn't be able to adjust the `require`. So I agree -- need to think about this more

> The `pack create-builder` command is used to copy hooks into the builder image via `builder.toml`. E.g.,

Reviewer 4: ```suggestion
The `pack builder create` command is used to copy hooks into the builder image via `builder.toml`. E.g.,
```

```toml
[[hooks]]
name = "app-hook"

> path = "./myhook"

Reviewer 4: Why do we need `path` if there's such a strong specification of the file path/name/extension (below)?

Reviewer 4: Ok, now I believe this is the path that's copied into the builder, at the path/filename defined below.

Reviewer 4: Alternative: Instead of exposing the "hook" concept to builder-image creators, we use the formats we support explicitly as tables under the base image. For example (assuming `[stack]` goes away):

```
[base-image.dockerfile]
path = "<string (path to file)>"
run = "<boolean>"
build = "<boolean>"
out = "<boolean>"
buildpacks = ["<string (buildpack ID)>"]
```

I'm open to other ideas that avoid introducing "hooks" to builder-image owners.

Reviewer 2: You lose flexibility. A hook can dynamically generate a Dockerfile. It could do that based on any information it has at the time it runs, such as the contents of the user's application or what buildpacks have passed detection. If you remove the hook then the builder can only add Dockerfiles that are set at the time the builder is created.

As an example, without the hook, you couldn't implement the [use case from the RFC](https://github.com/buildpacks/rfcs/pull/173/files#diff-aeaa4e3f3765f6b7ec295a8c92881bdeae1313f69fd338da209548883dafdf00R84-R93), where the hook takes the build and run Dockerfiles it generates from the app directory.

Reviewer 4: My example would dynamically generate a Dockerfile if `out = true`

Reviewer 4: Basically, I'm not trying to change the mechanism described below (hook.d is fine in the internals). It's how it manifests in the dev ex that concerns.

Reviewer 4: Actually, I question the value of `out = false` or a hook w/o `.out`. If we were to eliminate the static Dockerfile capability (i.e. if you want that go run `docker build` before `pack builder create`), then it might open up some more (and potentially simpler) UX options

Reviewer 2: Sorry, I'm not sure I follow. My present understanding is that like 90% of what's under [this section](https://github.com/buildpacks/rfcs/blob/dockerfiles/text/0000-dockerfiles.md#dynamically-applied-dockerfiles) is describing the internals. Most people wouldn't have to care about that, just the folks implementing this for tools like `pack`.

The actual user-facing part, that builder-builders would interact with is a.) the `[[hooks]]` table defined in the spec to indicate what hooks should be added to the builder image and b.) the hook itself.

Am I missing something?

>(Note: this format should not be considered final. Additional fields will be required to mark the format, target, etc. of the hook. These may be defined in subsequent spec PRs / sub-team RFCs.)

Or are you just trying to flush out this bit (ie. the format of `[[hooks]]`)? ^^

Reviewer 4: `[[hook]]` is a concept that a builder image owner would need to learn and understand it. Most likely there would be overlap with buildpack authors needing to understand it.

Reviewer 2: So is your concern with the name `hook`? Do you think it needs a different descriptive word?

The example you gave seems misleading to me, `[base-image.dockerfile]` implies that you'd have a static Dockerfile, not something that's executable. I think that was part of my confusion/question above. 

To me something like `[base-image.modifier]` would make more sense. It avoids the name hook. You could then have flags like `executable=true|false` and `format=dockerfile` (or whatever other formats might be supported in the future).

Reviewer 4: Yea, I admit my alternative is misleading (implies a static Dockerfile). But generally, I'm wondering if we can up-level the `format` such that the only construct you need to understand is the format of that artifact (i.e. the hook becomes an impl detail of how that thing is run).

> * <new lifecycle phase> run extensions' bin/generate, output Dockerfiles are written to a volume

Reviewer 5: Updated in f8e7bfa

> * <new lifecycle phase OR platform> apply Dockerfiles to build image

Reviewer 5: Updated in f8e7bfa

buildpacks = ["com.example.buildpack1"]
```

> (Note: this format should not be considered final. Additional fields will be required to mark the format, target, etc. of the hook. These may be defined in subsequent spec PRs / sub-team RFCs.)

Reviewer 3: We need as additional fields at least : 
- an `id`
- a description and/or url poiting to the project documenting the hook
to allow the builder to provide as information for the users what is the purpose of the hook, ...
and the : 
- `path` should support a URI: docker registry image or github project ...
- `name` should follow some rules to be unique, avoid to include non supported characters, ... (e.d. DNS fullt qualified name)

Each hook file path must be in the format `/cnb/hooks.d/<name>.(build.|run.|)<format>(.out|)`, where:

1. `build`, `run`, or empty specify the builder image, runtime base image, or both base images, respectively.
2. The only valid format is `Dockerfile`, although support for, e.g. LLB JSON, could be added in the future.
3. If the `.out` suffix is present and the file is executable, the hook is executed in the context of the app directory, and its output must match `<format>`.
4. If the `.out` suffix is not present, the contents of the file must match `<format>`.

Additionally, each hook file may be accompanied by an additional file of the same name, but with the suffix `.buildpacks`.
This file contains a newline-separated list of buildpack IDs.
If present, the hook is ignored unless one of the buildpacks in the list is selected during the buildpack detection phase.

> Hook files are evaluated (either read or executed) after the detection phase.

Reviewer 3: Can you include here a section to explain how the hooks will be selected from `/cnb/hooks.d/` ? If a group of hooks to be applied is generated, how will the hooks be sorted, .... ? Will the process rollback what it has been previously executed if a hook of the group fails now ?

Author: > If a group of hooks to be applied is generated, how will the hooks be sorted, .... ?

As specified,

> The resulting Dockerfiles are applied to the build-time and runtime base images in lexographical order by `<name>`.
(there are more details around that section: https://github.com/buildpacks/rfcs/pull/173/files#diff-aeaa4e3f3765f6b7ec295a8c92881bdeae1313f69fd338da209548883dafdf00R62)

> Will the process rollback what it has been previously executed if a hook of the group fails now ?

As specified,

> If an executable hook exits with a non-zero status value, the build fails.

Reviewer 4: I assume this is a new phase? Maybe the `extend` phase :)

All Dockerfiles are applied to base images after the hooks are evaluated and before the buildpack build phase.

> The resulting Dockerfiles are applied to the build-time and runtime base images in lexographical order by `<name>`.

Reviewer 4: Why not base the order on the order the hooks were defined in the` builder.toml? I think that would be more intuitive.

> Executable hook files must not modify the app directory when executed and may be evaluated in parallel.

Reviewer 15: @sclevine Most Paketo buildpacks use env variables to receive configuration from users. Many of these settings are not likely useful to hooks, but we do allow users to specify an env variable to pick versions of dependencies installed, ex: `BP_JVM_VERSION=8`. 

Will hooks have the same set of env variables set that you'd get during detect & build?

If an executable hook exits with a non-zero status value, the build fails.
If an executable hook exits with a zero status value and no output, the hook is ignored. 
Directories are ignored.
Files at the top-level of `/cnb/hooks.d/` that do not match the specified file name format result in a build failure.

Both Dockerfiles must accept `base_image` and `build_id` args.
The `base_image` arg allows the lifecycle to specify the original base image.
The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`. When the `$build_id` arg is referenced in a `RUN` instruction, all subsequent layerrs will be rebuilt on the next build (as the value will change).

A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before a runtime Dockerfile is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.

#### Example: App-specified Dockerfile Hook

> This example hook would allow an app to provide runtime and build-time base image extensions as "run.Dockerfile" and "build.Dockerfile."

Reviewer 3: This wording gives the impression that a user can provide their own build|run dockerfiles as you say `This example hook would allow an app to provide ....`. Is it correct ? If the answer is `True`, how then such `dockerfiles` will be included/mounted to the container when it will execute `pack build` ? Will they override the existing builder's image hooks ....

Author: This is just an example. A builder maintainer could create a hook that outputs run.Dockerfile and build.Dockerfile if present in the app directory. The hooks are executed in the context of the app directory, which is already in the container. Hooks are applied in lexicographical order, so in this example the app-specified Dockerfiles would be applied whenever the app.run.Dockerfile.out / app.build.Dockerfile.out hooks are applied.

Extensions generate Dockerfiles before the regular buildpack build phase.
To generate these Dockerfiles, the lifecycle executes the extension's `/bin/generate` executable with the same UID, GID, and interface as regular buildpacks.
However,
- Extensions `/bin/generate` must not write to the app directory.
- Extensions `<layers>` directory is replaced by an `<output>` directory.

> - If an extension is missing `/bin/generate`, the extension root is treated as a pre-populated `<output>` directory.

Reviewer: This combined with the following from above

> If an extension is missing `/bin/detect`, the extension root is treated as a pre-populated output directory (i.e., extensions can include a static build plan).

Makes the same directory the output of both detect and generate. Would it be tidier and more future proof to have two separate directories for these two purposes? 

The following is a proposed filetree for all possible files (although some of these would be mutually exclusive).
```
extension-root
├── bin
│   ├── detect
│   └── generate
├── detect
│   └── plan.toml
└── generate
    ├── Dockerfile
    ├── build.Dockerfile
    ├── build.toml
    ├── launch.toml
    └── run.Dockerfile
```

This organization might be particularly nice if we wanted to provide the same (static) options for regular buildpacks in the future.

Reviewer 5: I had the same thought myself, but I think it would be easier to implement this if both buildpacks and extensions followed the same convention - i.e., when we get to the point of providing static options for regular buildpacks, let's implement it for extensions also.

Reviewer 5: Updated in 3a26570

After `/bin/generate` executes, the `<output>` directory may contain
- `build.toml`, with the same contents as a normal buildpack's `build.toml`, but

>   - With an additional `args` table array with `name` and `value` fields that are provided as build args to `build.Dockerfile` or `Dockerfile`

Reviewer: Does this imply that extensions can set `unmet` plan entries, similar to buildpacks? I don't see a problem with that (it's probably a good idea) but wasn't sure if that was readily apparent.

Reviewer 5: Updated in f8e7bfa

```

#### Example: RPM Dockerfile Hook (app-based)

> This example hook would allow a builder to install RPMs for each language runtime, based on the app directory.

Reviewer 3: This wording is not clear for me. Who is the `builder` ? What `app-based` means and what is the different with the next scenario where you say `... (buildpack-based)` ?

Author: builder = builder image
"based on the app directory" = based on the contents of the app directory
buildpack-based = only applied for a particular selection of buildpacks

Happy to clarify the wording in the RFC

>   - Without the `slices` table array

Reviewer: The only thing not excluded here is `labels`. It seems kinda weird to set labels this way when we also have the Dockerfile interface? Wdyt?

If we don't allow labels zero of the existing launch.toml fields are allowed which really makes me wonder why we are trying to call this the same file. Maybe we could just have a new `run.toml`? That name has a nice symmetry with `run.Dockerfile` (like `build.toml` and `build.Dockerfile`).

Reviewer 5: Updated in f8e7bfa

Note: The Dockerfiles referenced must disable rebasing, and build times will be slower compared to buildpack-provided runtimes.

##### `/cnb/hooks.d/app.Dockerfile.out`
```
#!/bin/sh
[[ -f Gemfile.lock ]] && cat /cnb/hooks.d/app.Dockerfile.d/Dockerfile-ruby

> [[ -f package.json ]] && cat /cnb/hooks.d/app.Dockerfile.d/Dockerfile-node

Reviewer 3: As such checks (to detect if a language/runtime is present within the app's project to be build) has already been done during the `detect` phase, why don't we simply pass a METADATA, env var or yaml/toml file to the hook executable instead of doing again the same check ?

Author: > has already been done during the detect phase

As mentioned above, that would couple buildpacks to hooks. Some buildpacks would become entirely dependent on hooks. For a given app and set of buildpacks, it should be possible to create a working builder that doesn't require hooks (since hooks will be slower and break rebase in many cases)

Reviewer 3: > As mentioned above, that would couple buildpacks to hooks. Some buildpacks would become entirely dependent on hooks.

No. As the hook could rely on the METADATA - https://github.com/buildpacks/spec/blob/main/platform.md#detector coming from the `detect` phase, I dont see why you say that we will create a strong dependency with the `build` phase to be executed of a buildpack. such METADATA could simply provide as additional information: JDK11, .... and then the hook will calculate what it should include within the run and/or build Dockerfile generated

Author: That metadata (`require`s) has to come from other buildpacks -- that's the coupling I'm worried about.

That said, I'm coming around to this idea, given the observation here: https://github.com/buildpacks/rfcs/pull/173#discussion_r675625620

I'll propose some changes for next week, appreciate the feedback :)

> If no Dockerfiles are present, `/bin/generate` may still consume build plan entries and add metadata to `build.toml`/`launch.toml`.
> 

Reviewer: ```suggestion
If no Dockerfiles are present, `/bin/generate` may still consume build plan entries.
```

I think this metadata reference is left over from the days when the bom was in the build/launch files? There is no way to set metadata in the current [buildpack spec](https://github.com/buildpacks/spec/blob/main/buildpack.md#buildtoml-toml).

Reviewer 5: Updated in f8e7bfa

Dockerfiles are applied to their corresponding base images after all extensions are executed and before any regular buildpacks are executed.
Dockerfiles are applied in the order determined during buildpack detection. When multiple Dockerfiles are applied, the intermediate image generated from the application of the current Dockerfile will be provided as the `base_image` ARG to the next Dockerfile. Dockerfiles that target the run image (only) may ignore the provided `base_image` (e.g., `FROM some-other-image`). Dockerfiles that change the runtime base image may still use `COPY --from=${base_image}`.

All Dockerfiles are provided with `base_image` and `build_id` args.
The `base_image` arg allows the Dockerfile to reference the original base image.

> The `build_id` arg allows the Dockerfile to invalidate the cache after a certain layer and must be defaulted to `0`. The executor of the Dockerfile will provide the `build_id` as a UUID (this eliminates the need to track this variable). 

Reviewer: ```suggestion
The `build_id` arg allows the Dockerfile to invalidate the cache after a certain layer. The executor of the Dockerfile will provide the `build_id` as a UUID (this eliminates the need to track this variable). 
```

Does it matter what it is defaulted to if the executor always sets it? Seems like an unnecessary validation point for extension authors to get wrong.

Reviewer 5: I had the same thought here: https://github.com/buildpacks/rfcs/pull/173#discussion_r903807205

Reviewer 5: Updated in 3a26570

RUN yum install ruby
...
```

##### `/cnb/hooks.d/ruby.Dockerfile.buildpacks`
```
com.example.ruby
```

##### `/cnb/hooks.d/node.Dockerfile`
```
...
LABEL io.buildpacks.rebasable=false # default, not needed
RUN yum install nodejs
...
```

##### `/cnb/hooks.d/node.Dockerfile.buildpacks`
```
com.example.node
```

### Dockerfiles for Base Images

The same Dockerfile format may be used to create new base images or modify existing base images outside of the app build process (e.g., before creating a builder). Any specified labels override existing values.

Dockerfiles that are used to create a base image must create a `/cnb/image/genpkgs` executable that outputs a [CycloneDX](https://cyclonedx.org)-formatted list of packages in the image with PURL IDs when invoked. This executable is executed after all Dockerfiles are applied, and the output replaces the label `io.buildpacks.sbom`. This label doubles as a Software Bill-of-Materials for the base image. In the future, this label will serve as a starting point for the application SBoM.

### Example Dockerfiles

Dockerfile used to create a runtime base image:

```
ARG base_image
FROM ${base_image}

> ARG build_id=0

Reviewer 9: Are these automatically added? Or does a hook need to output dockerfiles that contain this for each dockerfile it outputs?

Author: Hooks must accept the two args.

Author: https://github.com/buildpacks/rfcs/pull/173/files#diff-aeaa4e3f3765f6b7ec295a8c92881bdeae1313f69fd338da209548883dafdf00R72

Author: It appears that (at least with BuildKit) Docker no longer complains about extra build args.
So these are totally optional  :)

LABEL io.buildpacks.image.distro=ubuntu
LABEL io.buildpacks.image.version=18.04

> LABEL io.buildpacks.rebasable=true

Author: I only intended it to mean that `pack rebase` is allowed without `--force`.
Are you worried about a specific use case?

> A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically on the base image before a runtime Dockerfile is applied and must be explicitly set to `true` if desired. If multiple Dockerfiles are applied, all must set `io.buildpacks.rebasable=true` for the final value to be `true`. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.

Reviewer: If I am understanding correctly a rebasing an extended app image would treat extension layers like app layers rather than like run-image layers.

How does the run image top-layer get set correctly in [`io.buildpacks.lifecycle.metadata`](https://github.com/buildpacks/spec/blob/main/platform.md#iobuildpackslifecyclemetadata-json) so that we can rebase the correct layers? If we simply passed the extended run image to export the extension layers would lost, given how this works currently.

If an extension entire switches the run image (rather than just extending the existing one) how does the stack metadata get set correctly on the app image so that new versions can be found?

Reviewer 5: Discussed in implementation sync:

This isn't relevant for "phase 1", but in the future, when we allow the run image to be extended:
* The `analyzer` when resolving creds for the run image can record the top layer in analyzed.toml
* The `exporter` can check that the top layer in analyzed.toml is a layer in the run image it was provided
* If the top layer is found (the run image was extended, not switched), then
  * `io.buildpacks.lifecycle.metadata` `runImage.topLayer` and `runImage.reference` can be derived from analyzed.toml
* If the top layer is not found (the run image was switched, and possibly also extended), then
  * The `extender` when applying Dockerfiles needs to keep track of the top layer for each image, so that it can know when a base image has been switched; the extender will provide the top layer of the run image (prior to any extensions) in some sort of output file OR it can add a label to the extended run image e.g., `io.buildpacks.rebase.metadata`
  * `io.buildpacks.lifecycle.metadata` `runImage.topLayer` and `runImage.reference` can be derived from the output of the extender

Having typed all of that out, maybe we need to be looking at ChainID (vs DiffID) in determining whether an image was switched or extended. But the principle is the same.

cc @ekcasey

Reviewer 5: Just to note - all of this is moot if `io.buildpacks.rebasable` is false. Then the top layer will just be the top layer in the run image that was provided to export.

Reviewer 5: Updated in 3a26570

> ```

Reviewer: Given we are allowing static Dockerfiles in the case where there is no `bin/generate` script, is this example misleading? Should this instead be an example w/o a generate executable?

RUN groupadd cnb --gid ${CNB_GROUP_ID} && \
  useradd --uid ${CNB_USER_ID} --gid ${CNB_GROUP_ID} -m -s /bin/bash cnb

USER ${CNB_USER_ID}:${CNB_GROUP_ID}

COPY genpkgs /cnb/image/genpkgs
```

`run.Dockerfile` for use with the example `app.run.Dockerfile.out` hook that always installs the latest version of curl:
```
ARG base_image
FROM ${base_image}
ARG build_id=0

RUN echo ${build_id}

RUN apt-get update && apt-get install -y curl && rm -rf /var/lib/apt/lists/*
```
(note: this Dockerfile disables rebasing, as OS package installation is not rebasable)

`run.Dockerfile` for use with the example `app.run.Dockerfile.out` hook that installs a special package to /opt:
```
ARG base_image
FROM ${base_image}
ARG build_id=0

LABEL io.buildpacks.rebasable=true

RUN curl -L https://example.com/mypkg-install | sh # installs to /opt
```
(note: rebasing is explicitly allowed because only a single directory in /opt is created)

### Future: Dynamic Run-time Base Image Selection

The outcomes targeted by https://github.com/buildpacks/rfcs/pull/174 could be achieved with an additional type of hook.

For example, we could introduce a `ref` format that replaces the image instead of extending it:
```

> /cnb/hooks.d/myhook.run.ref # contains: index.docker.io/example/my-ruby-run-image

Reviewer 9: This could simple be achieved by using

```dockerfile
FROM index.docker.io/example/my-ruby-run-image
```
instead?

Reviewer 9: And instead of ref we could use a normal dockerfile hook?

Author: I originally disliked this idea because I assumed that `docker build` would fail on unused build args. Given that it does not complain about them, I agree this is sufficient.

> /cnb/hooks.d/myhook.run.buildpacks # contains: com.example.ruby-buildpack

Author: Is it less confusing if we take @samj1912's suggestion and only allow this feature via a FROM-only Dockerfile?

```

This would simplify buildpacks by removing the need for PURLs outside of SBoM generation. The `packages` table could be removed from both buildpack.toml and `/bin/detect`.

# Drawbacks

> [drawbacks]: #drawbacks

Reviewer 4: It's unfortunate that this proposal introduces a new concept (hooks) that a builder-image maintainer needs to learn; especially when this and #172 begin to simplify and remove some concepts. I also work that the hook concept is not very intuitive.

Reviewer 2: >I also work that the hook concept is not very intuitive.

This is subjective, can you qualify this statement with some details or an example? 

I would argue that you find the concept of a "hook" (or pre/post script, before/after, wrapper, etc..) in tons of places in the computer world. The general concept is something many sys-admins and dev's have probably already seen.

I agree it adds some complexity because to use a hook, you need to understand the expected output format & what information you have available to generate your output (as well as basic scripting). At the same time, it's optional. You don't need any hooks to generate a builder, so a builder-builder doesn't really need to know this unless they are trying to do more customizations, which you could say is a more advanced task.

My $0.02 only, but it seems like the ratio of complexity added versus additional functionality for users is quite reasonable.

Reviewer 4: Here is what I believe makes it not intuitive:

* It isn't clear when the "hook" runs (before all buildpacks? before each buildpack? after buildpacks?)
* it isn't obvious that the filename defines the format and thus what formats are supported
* It isn't clear that the "name" determines the order
* It isn't clear what happens when there's more than one hook
* Generally, I don't think it's obvious what the scope is (what are the things a hook can do? Can it mutate a buildpack?)

These are not very difficult things to learn, but it's yet another thing to learn (which is a shame given that we're looking to remove things like the stack and mixin to make buildpacks easier to grok). 

By comparison to your examples, a pre/post script usually stems from the same construct as the thing it's wrapping around (another executable). Similarly, the pre/post proposal for project.toml adds buildpacks before/after buildpacks. These things are all homogenous.  But here, we have a concept that introduces a new mechanism into the existing "order". So now we have heterogeneous things running in some order. I think that's what makes it difficult to reason about.

Reviewer 2: OK, so there seem to be some specifics in the RFC that you think need to be clarified (I agree with some of these btw). IMHO, that seems like something that can be addressed and not a fundamental flaw in the way this is working. Is that your thought as well?

-------

I will have to respectfully disagree on the "hook" as a concept though. There's is a fairly basic lifecycle that buildpacks flow through, you're talking about being able to inject some actions at a couple of points in that lifecycle. This happens all the time in software, they even created AOP to formalize the concept. It happens in software systems as well, the venerable HTTP Servlet Filter in the Java world is an example. As long as we document and have a picture of the lifecycle & indicate where the hooks execute (which is why I agree we need to clarify that point), then we should be fine. 

I also strongly believe the complexity we are adding is more than worth it with the functionality this is adding & the number of problems this can solve.

Reviewer 2: Sure. No disagreement there. Learning nothing is simpler than having to learn something, even if it's simple. 

What I'm left trying to understand is this. If your [bullet list of concerns](https://github.com/buildpacks/rfcs/pull/173#discussion_r676041016) were addressed then would the complexity that this RFC adds be OK? or is it your opinion that this RFC is fundamentally too complex and not something we should add?

Reviewer 4: No, I'm generally in favor of this RFC. I'm only trying to push us to find an abstraction over the hook mechanism that makes it's easier for users to understand (especially buildpack authors and builder-image owners).

Reviewer 3: I'm also in favour of this RFC. Nevertheless, It is still very hard (as reported too by Joe) to understand different very important points such as : 
- When and how Hook(s) will be executed ?
- How a hook will become part by example of the build execution A and not B, ... ?, 
- The parameters that a hook will consume (= IN) and what they will generate (= OUT),
- How the buildpacks will react if a Docker build fails (return code, error message, ...) ?,
- Where the new layers created for a base image or run image will be stored ?
- What will be the file permissions defined for such new files added to a layer (as normally root user will be used to execute them) ?
- If the RFC only populates dynamically Dockerfiles OR could consume Dockerfiles provided by a developer under which path of their application currently developed ?

**Remark**: That should be great to have a simple text workflow explaining how hooks will be integrated within the existing buildpacks phases (e.g: DETECTION --> ANALYSIS --> BUILD --> EXPORT ==> DETECTION --> ANALYSIS --> HOOK --> BUILD --> EXPORT)

Author: > When and how Hook(s) will be executed ?
How a hook will become part by example of the build execution A and not B, ... ?,
The parameters that a hook will consume (= IN) and what they will generate (= OUT),
How the buildpacks will react if a Docker build fails (return code, error message, ...) ?,

These are covered in the RFC. Can you be more specific about what details you feel are missing?

> Where the new layers created for a base image or run image will be stored ?

Do you mean stored as in cached? Caching strategy would be up to the platform to decide.

> What will be the file permissions defined for such new files added to a layer (as normally root user will be used to execute them) ?

The Dockerfile would run as root on the base image and produce a new image. No other changes to file permissions would occur.

> If the RFC only populates dynamically Dockerfiles OR could consume Dockerfiles provided by a developer under which path of their application currently developed ?

As specified, buildpacks could create Dockerfiles in their ARGV[1] directory. They could copy them from the app directory, as described in the example, but that's not built-in to the API.

Reviewer 5: > buildpacks could create Dockerfiles

I think you mean hooks?

# How it Works
[how-it-works]: #how-it-works

Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 

### Builder-specified Dockerfiles

> A builder may specify any number of executable "hooks" in `/cnb/hooks.d/`.

Author: Builder refers to a builder image (output of `pack create-builder`).
Executable is an executable file (e.g., binary, script, etc.).
`pack create-builder` would be responsible for installing hooks.

Will add these details to the RFC.

> Hook files in this directory are executed in the context of app directory and must output Dockerfiles that are applied to the build-time and runtime base images before an image is built.

Author: A new phase would likely need to be introduced before detection.

> Hook files in this directory are executed in the context of app directory and must output Dockerfiles that are applied to the build-time and runtime base images before an image is built.

Reviewer 3: Can you then add it to your last review (See my review of today) as that was not at all mentioned like also what will be the Inputs/Outputs (e.g See what it is described as example part of the spec for the builder - https://github.com/buildpacks/spec/blob/main/platform.md#builder) ?

If a hook exits with a non-zero status value, the build fails. If a hook exits with a zero status value and no output, the hook is ignored. Directories and non-executable files are ignored.

> Each executable must be in the format `/cnb/hooks.d/<name>.(build.|run.|)<format>`, where build, run, or empty specify the build-time base image, runtime base image, or both bash images, respectively. The only valid format is `Dockerfile`, although support for, e.g. LLB JSON, could be added in the future. 

Author: Will clarify the wording for (1).

For (2) and (3), there's no connection between the hooks and buildpacks. The names are up to the bulider author.

For (4), as stated, the file name would be `/cnb/hooks.d/<name>.(build.|run.|)<format>`, with a descriptive name decided by the buildpack author.

A runtime Dockerfile is applied to the selected runtime base image after the detection phase.
A build-time Dockerfile is applied to the build-time base image before the detection phase.

Both Dockerfiles must accept `base_image` and `build_id` args.

> The `base_image` arg allows the lifecycle to specify the original base image.

Reviewer 16: Are we allowed to override/ignore this `base_image` and specify our own?

Author: Anything is valid, as long as the Dockerfile accepts the ARG and the resulting image is spec-compliant.

The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`. When the `$build_id` arg is referenced in a `RUN` instruction, all subsequent layerrs will be rebuilt on the next build (as the value will change).

A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of app-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before `run.Dockerfile` is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.

#### Example: App-specified Dockerfile Hook

> This example hook would allow an app to provide runtime and build-time base image extensions as "run.Dockerfile" and "build.Dockerfile."

Author: These Dockerfiles are just examples that are read by the hooks below -- they are not Builder-specified Dockerfiles.

The app developer can decide whether the extensions are rebasable.

##### `/cnb/hooks.d/app.build.Dockerfile`
```
#!/bin/sh
cat build.Dockerfile
```
##### `/cnb/hooks.d/app.run.Dockerfile`
```
#!/bin/sh
cat run.Dockerfile
```

#### Example: RPM Dockerfile Hook

This example hook would allow a builder to install RPMs for each language runtime.

> Note: The Dockerfiles referenced must disable rebasing, and build times will be slower compared to buildpack-provided runtimes.

Reviewer 3: Can you at the beginning of this RFC describe clearly if rebasing will be supported or not when dockerfiles have been provided by an application (= developer project) or by the platform ?

If run images cannot be rebased, what will be then in this case the workaround (if it exists) as rebasing is a key feature ?

Author: As mentioned in the first section:
> A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before a runtime Dockerfile is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.

So it's up to the Dockerfile author to enable rebasing. That said, Dockerfiles installing OS packages (like RPMs) should never allow rebasing, because it is unsafe. There is no workaround for this other than not using OS packages at build-time. OS packages are arbitrary functions that can modify existing files.

Reviewer 16: > OS packages are arbitrary functions that can modify existing files.

Is that the only criterium for knowing if a Dockerfile would be rebasable? So if you knew for sure that your RPM didn't change any files but that it only added files you could safely set the rebasable flag to true?

Author: If installing your RPM (and all dependencies of that RPM) only adds net-new files, and if those files are not present in the base image (or any future base image), then it's likely safe to mark the image as rebasable.

That said, unless you unpack the RPM manually, installing your RPM modifies the package DB in /var/lib/rpm. Rebase would still be unsafe and could also, e.g., cause security scanners to pick up the wrong package versions after a rebase.

Reviewer 5: Marking as resolved, but please feel free to reopen if needed.

##### `/cnb/hooks.d/app.Dockerfile`
```
#!/bin/sh
[[ -f Gemfile.lock ]] && cat /cnb/hooks.d/app.Dockerfile.d/Dockerfile-ruby
[[ -f package.json ]] && cat /cnb/hooks.d/app.Dockerfile.d/Dockerfile-node
```

> ### Platform-specified Dockerfiles

Reviewer 3: Can you please describe what is the `platform-` referenced here vs `app-`  or `builder-` ?
Is an application a developer project ? If yes, how will it be possible to copy the files created locally to the `/cnb/hook.d` folder ?

How hooks will be defined part of a buildpacks or platform ?

Author: The section just describes how to use Dockerfiles to extend a stack image. It's not related to hooks. Will clarify.

> How hooks will be defined part of a buildpacks or platform ?

Hooks will be added to a builder as part of `pack create-builder`.
